{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "usage: ipykernel_launcher.py [-h] [--header] input_file [output_file]\n",
      "ipykernel_launcher.py: error: unrecognized arguments: -f\n"
     ]
    },
    {
     "ename": "SystemExit",
     "evalue": "2",
     "output_type": "error",
     "traceback": [
      "An exception has occurred, use %tb to see the full traceback.\n",
      "\u001b[0;31mSystemExit\u001b[0m\u001b[0;31m:\u001b[0m 2\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "def uniqMaxCol(df=None,fname=None,idCol=8, maxCol=13):\n",
    "    ''' Keep the record among duplicates by maxCol\n",
    "    '''\n",
    "    if df is None:\n",
    "        df = pd.read_table(fname,header=None,dtype=default_dtype)\n",
    "    dup = df[idCol].duplicated(keep=False)\n",
    "    ddup,duniq = df.loc[dup],df.loc[~dup]\n",
    "    ddup[maxCol].astype('float16',copy=False)\n",
    "#     ddup[maxCol]=ddup[maxCol].astype('float16')\n",
    "    def maxBycol( val,col=maxCol):\n",
    "        if len(val)>1:\n",
    "            signal = val[col]\n",
    "            ix = signal.idxmax()\n",
    "            return val.loc[[ix]]\n",
    "        else:\n",
    "            return val\n",
    "    ddedup = ddup.groupby(idCol,as_index=0).apply(maxBycol)\n",
    "    dout = duniq.append(ddedup)\n",
    "    dout.columns = default_header\n",
    "    return dout\n",
    "\n",
    "import sys\n",
    "lst='''feat_chrom,str\n",
    "feat_source,str\n",
    "feat_type,str\n",
    "feat_start,int\n",
    "feat_end,int\n",
    "feat_score,str\n",
    "feat_strand,str\n",
    "feat_phase,str\n",
    "feat_attributes,str\n",
    "peak_chrom,str\n",
    "peak_start,int\n",
    "peak_end,int\n",
    "peak_id,str\n",
    "peak_score,str\n",
    "peak_strand,str\n",
    "peak_fc,str\n",
    "peak_logP,str\n",
    "peak_logQ,str\n",
    "peak_sumLoc,str'''.splitlines()\n",
    "default_header,default_dtype = zip(*[x.split(',') for x in lst])\n",
    "\n",
    "default_dtype={i:d for i,d in enumerate(default_dtype)}\n",
    "if __name__=='__main__':\n",
    "    import argparse\n",
    "\n",
    "    parser = argparse.ArgumentParser(description='Keep the record among duplicates by maxCol')\n",
    "#     parser.add_argument('integers', metavar='N', type=int, nargs='+',\n",
    "#                         help='an integer for the accumulator')\n",
    "    parser.add_argument('fname', metavar='input_file', type=str, nargs=1,\n",
    "                        help='Input file')\n",
    "    parser.add_argument('ofname', metavar='output_file', type=str, nargs='?', default=None,\n",
    "                        help='Output file')\n",
    "    #     parser.add_argument('ofname', metavar='input_file', type=str, nargs='1',\n",
    "#                         help='Input file')\n",
    "    parser.add_argument('--header', dest='incHeader', action='store_true',\n",
    "                        help='include the header',default=None)\n",
    "    args = parser.parse_args()\n",
    "#     print args\n",
    "#     print args.ofname\n",
    "#     fname = sys.argv[1]\n",
    "#     ofname= sys.argv[2:]\n",
    "    df = uniqMaxCol(fname=args.fname[0],idCol=8,maxCol=13)\n",
    "    \n",
    "#     saver = lambda df,*args: \n",
    "    df.to_csv(\n",
    "        path_or_buf=args.ofname or sys.stdout,\n",
    "        sep='\\t',index=0,\n",
    "        header=args.incHeader)  \n",
    "#     saver(df,args.ofname or sys.stdout)\n",
    "#         df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exit(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[NbConvertApp] Converting notebook uniqMaxCol.ipynb to python\n",
      "[NbConvertApp] Writing 7556 bytes to uniqMaxCol.py\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "jupyter nbconvert --to python uniqMaxCol.ipynb\n",
    "sed -i \"1i #!/usr/bin/env python2\" *.py\n",
    "chmod +x uniqMaxCol.py\n",
    "\n",
    "cp *.py $repos/BrachyPhoton/pipeline_rnaseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bd1\tphytozomev10\tgene\t10581\t11638\t.\t+\t.\tID=Bradi1g00200.v3.1;Name=Bradi1g00200;ancestorIdentifier=Bradi1g00200.v2.1\tBd1\t9738\t10049\tNA_peak_2\t45\t.\t5.63936\t7.79234\t4.56778\t20\n",
      "Bd1\tphytozomev10\tgene\t18453\t21742\t.\t+\t.\tID=Bradi1g00210.v3.1;Name=Bradi1g00210;ancestorIdentifier=Bradi1g00210.v2.1\tBd1\t9738\t10049\tNA_peak_2\t45\t.\t5.63936\t7.79234\t4.56778\t20\n",
      "\n",
      "\n",
      "feat_chrom\tfeat_source\tfeat_type\tfeat_start\tfeat_end\tfeat_score\tfeat_strand\tfeat_phase\tfeat_attributes\tpeak_chrom\tpeak_start\tpeak_end\tpeak_id\tpeak_score\tpeak_strand\tpeak_fc\tpeak_logP\tpeak_logQ\tpeak_sumLoc\n",
      "Bd1\tphytozomev10\tgene\t10581\t11638\t.\t+\t.\tID=Bradi1g00200.v3.1;Name=Bradi1g00200;ancestorIdentifier=Bradi1g00200.v2.1\tBd1\t9738\t10049\tNA_peak_2\t45\t.\t5.63936\t7.79234\t4.56778\t20\n",
      "Bd1\tphytozomev10\tgene\t18453\t21742\t.\t+\t.\tID=Bradi1g00210.v3.1;Name=Bradi1g00210;ancestorIdentifier=Bradi1g00210.v2.1\tBd1\t9738\t10049\tNA_peak_2\t45\t.\t5.63936\t7.79234\t4.56778\t20\n",
      "Traceback (most recent call last):\n",
      "  File \"./uniqMaxCol.py\", line 78, in <module>\n",
      "    header=args.incHeader)  \n",
      "  File \"/usr/lib/python2.7/dist-packages/pandas/core/frame.py\", line 1332, in to_csv\n",
      "    formatter.save()\n",
      "  File \"/usr/lib/python2.7/dist-packages/pandas/core/format.py\", line 1506, in save\n",
      "    self._save()\n",
      "  File \"/usr/lib/python2.7/dist-packages/pandas/core/format.py\", line 1606, in _save\n",
      "    self._save_chunk(start_i, end_i)\n",
      "  File \"/usr/lib/python2.7/dist-packages/pandas/core/format.py\", line 1633, in _save_chunk\n",
      "    lib.write_csv_rows(self.data, ix, self.nlevels, self.cols, self.writer)\n",
      "  File \"pandas/lib.pyx\", line 1073, in pandas.lib.write_csv_rows (pandas/lib.c:19757)\n",
      "IOError: [Errno 32] Broken pipe\n",
      "\n",
      "\n",
      "Bd1\tphytozomev10\tgene\t10581\t11638\t.\t+\t.\tID=Bradi1g00200.v3.1;Name=Bradi1g00200;ancestorIdentifier=Bradi1g00200.v2.1\tBd1\t9738\t10049\tNA_peak_2\t45\t.\t5.63936\t7.79234\t4.56778\t20\n",
      "Bd1\tphytozomev10\tgene\t18453\t21742\t.\t+\t.\tID=Bradi1g00210.v3.1;Name=Bradi1g00210;ancestorIdentifier=Bradi1g00210.v2.1\tBd1\t9738\t10049\tNA_peak_2\t45\t.\t5.63936\t7.79234\t4.56778\t20\n"
     ]
    }
   ],
   "source": [
    "! rm -f tmp_dedup.bed\n",
    "! ./uniqMaxCol.py tmp.bed >tmp_dedup.bed\n",
    "! head tmp_dedup.bed -n2\n",
    "! echo \\\\n\n",
    "! ./uniqMaxCol.py tmp.bed --header |head -n3\n",
    "! echo \\\\n\n",
    "! ./uniqMaxCol.py tmp.bed tmp_dedup.bed && head tmp_dedup.bed -n2\n",
    "# ! head tmp_dedup.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feat_chrom</th>\n",
       "      <th>feat_source</th>\n",
       "      <th>feat_type</th>\n",
       "      <th>feat_start</th>\n",
       "      <th>feat_end</th>\n",
       "      <th>feat_score</th>\n",
       "      <th>feat_strand</th>\n",
       "      <th>feat_phase</th>\n",
       "      <th>feat_attributes</th>\n",
       "      <th>peak_chrom</th>\n",
       "      <th>peak_start</th>\n",
       "      <th>peak_end</th>\n",
       "      <th>peak_id</th>\n",
       "      <th>peak_score</th>\n",
       "      <th>peak_strand</th>\n",
       "      <th>peak_fc</th>\n",
       "      <th>peak_logP</th>\n",
       "      <th>peak_logQ</th>\n",
       "      <th>peak_sumLoc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Bd1</td>\n",
       "      <td>phytozomev10</td>\n",
       "      <td>gene</td>\n",
       "      <td>10581</td>\n",
       "      <td>11638</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=Bradi1g00200.v3.1;Name=Bradi1g00200;ancesto...</td>\n",
       "      <td>Bd1</td>\n",
       "      <td>9738</td>\n",
       "      <td>10049</td>\n",
       "      <td>NA_peak_2</td>\n",
       "      <td>45</td>\n",
       "      <td>.</td>\n",
       "      <td>5.63936</td>\n",
       "      <td>7.79234</td>\n",
       "      <td>4.56778</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bd1</td>\n",
       "      <td>phytozomev10</td>\n",
       "      <td>gene</td>\n",
       "      <td>18453</td>\n",
       "      <td>21742</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=Bradi1g00210.v3.1;Name=Bradi1g00210;ancesto...</td>\n",
       "      <td>Bd1</td>\n",
       "      <td>9738</td>\n",
       "      <td>10049</td>\n",
       "      <td>NA_peak_2</td>\n",
       "      <td>45</td>\n",
       "      <td>.</td>\n",
       "      <td>5.63936</td>\n",
       "      <td>7.79234</td>\n",
       "      <td>4.56778</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bd1</td>\n",
       "      <td>phytozomev10</td>\n",
       "      <td>gene</td>\n",
       "      <td>28710</td>\n",
       "      <td>31462</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=Bradi1g00213.v3.1;Name=Bradi1g00213;ancesto...</td>\n",
       "      <td>Bd1</td>\n",
       "      <td>32587</td>\n",
       "      <td>33418</td>\n",
       "      <td>NA_peak_3</td>\n",
       "      <td>138</td>\n",
       "      <td>.</td>\n",
       "      <td>7.37483</td>\n",
       "      <td>17.6728</td>\n",
       "      <td>13.8093</td>\n",
       "      <td>292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bd1</td>\n",
       "      <td>phytozomev10</td>\n",
       "      <td>gene</td>\n",
       "      <td>47522</td>\n",
       "      <td>49251</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=Bradi1g00215.v3.1;Name=Bradi1g00215</td>\n",
       "      <td>Bd1</td>\n",
       "      <td>33530</td>\n",
       "      <td>33883</td>\n",
       "      <td>NA_peak_4</td>\n",
       "      <td>59</td>\n",
       "      <td>.</td>\n",
       "      <td>4.95704</td>\n",
       "      <td>9.32209</td>\n",
       "      <td>5.98403</td>\n",
       "      <td>197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bd1</td>\n",
       "      <td>phytozomev10</td>\n",
       "      <td>gene</td>\n",
       "      <td>52489</td>\n",
       "      <td>52869</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=Bradi1g00217.v3.1;Name=Bradi1g00217</td>\n",
       "      <td>Bd1</td>\n",
       "      <td>33530</td>\n",
       "      <td>33883</td>\n",
       "      <td>NA_peak_4</td>\n",
       "      <td>59</td>\n",
       "      <td>.</td>\n",
       "      <td>4.95704</td>\n",
       "      <td>9.32209</td>\n",
       "      <td>5.98403</td>\n",
       "      <td>197</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  feat_chrom   feat_source feat_type  feat_start  feat_end feat_score  \\\n",
       "0        Bd1  phytozomev10      gene       10581     11638          .   \n",
       "1        Bd1  phytozomev10      gene       18453     21742          .   \n",
       "2        Bd1  phytozomev10      gene       28710     31462          .   \n",
       "3        Bd1  phytozomev10      gene       47522     49251          .   \n",
       "4        Bd1  phytozomev10      gene       52489     52869          .   \n",
       "\n",
       "  feat_strand feat_phase                                    feat_attributes  \\\n",
       "0           +          .  ID=Bradi1g00200.v3.1;Name=Bradi1g00200;ancesto...   \n",
       "1           +          .  ID=Bradi1g00210.v3.1;Name=Bradi1g00210;ancesto...   \n",
       "2           -          .  ID=Bradi1g00213.v3.1;Name=Bradi1g00213;ancesto...   \n",
       "3           +          .             ID=Bradi1g00215.v3.1;Name=Bradi1g00215   \n",
       "4           +          .             ID=Bradi1g00217.v3.1;Name=Bradi1g00217   \n",
       "\n",
       "  peak_chrom  peak_start  peak_end    peak_id peak_score peak_strand  peak_fc  \\\n",
       "0        Bd1        9738     10049  NA_peak_2         45           .  5.63936   \n",
       "1        Bd1        9738     10049  NA_peak_2         45           .  5.63936   \n",
       "2        Bd1       32587     33418  NA_peak_3        138           .  7.37483   \n",
       "3        Bd1       33530     33883  NA_peak_4         59           .  4.95704   \n",
       "4        Bd1       33530     33883  NA_peak_4         59           .  4.95704   \n",
       "\n",
       "  peak_logP peak_logQ peak_sumLoc  \n",
       "0   7.79234   4.56778          20  \n",
       "1   7.79234   4.56778          20  \n",
       "2   17.6728   13.8093         292  \n",
       "3   9.32209   5.98403         197  \n",
       "4   9.32209   5.98403         197  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = uniqMaxCol(fname='tmp.bed')\n",
    "df.head(5)\n",
    "# print saver(df)[:300]\n",
    "# df.to_csv('tmp_dedup.bed',sep='\\t',index=0,header=None)\n",
    "# !head tmp_dedup.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "</p><p><b>Column 1:</b> \"seqid\"\n",
      "<p><b>Column 2:</b> \"source\"\n",
      "<p><b>Column 3:</b> \"type\"\n",
      "<p><b>Column 6:</b> \"score\"\n",
      "<p><b>Column 7:</b> \"strand\"\n",
      "<p><b>Column 8:</b> \"phase\"\n",
      "<p><b>Column 9:</b> \"attributes\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "\r",
      "  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\r",
      "  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\r",
      "100 49544    0 49544    0     0   103k      0 --:--:-- --:--:-- --:--:--  103k\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "curl http://gmod.org/wiki/GFF3 | grep -E \"Column [0-9]:\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Script is being sourced\n",
      "$BASH_SOURCE /home/feng/envs/ref/bin/activate\n",
      "/home/feng/envs/ref/bin\n",
      "[SOURCE]ing [UTILITIES] from /home/feng/envs/ref/bin/util.sh\n",
      "Adding /home/feng/envs/ref to PATH\n",
      "A=/home/feng/ref/Brachypodium_Bd21_v3-1/sequence/Bowtie2Index/Bdistachyon314_Bd.1.bt2\n",
      "[Test] GTF=/home/feng/ref/Brachypodium_Bd21_v3-1/annotation/Bdistachyon_314_v3.1.gene_exons.gtf\n",
      "[Test] GFF=/home/feng/ref/Brachypodium_Bd21_v3-1/annotation/Bdistachyon_314_v3.1.gene_exons.gff3\n",
      "[Test] GSIZE=/home/feng/ref/Brachypodium_Bd21_v3-1/genome.sizes\n",
      "[Test] FA_ADAPTER=/home/Program_NGS_sl-pw-srv01/Trimmomatic-0.32/adapters/TruSeq3-SE.fa\n",
      "[Test] REF=/home/feng/ref/Brachypodium_Bd21_v3-1\n",
      "[Test] IDX_BOWTIE2=/home/feng/ref/Brachypodium_Bd21_v3-1/sequence/Bowtie2Index/Bdistachyon314_Bd\n",
      "lrwxrwxrwx 1 feng feng 75 Jun  8 12:15 /home/feng/ref/Brachypodium_Bd21_v3-1/sequence/Bowtie2Index/Bdistachyon314_Bd.1.bt2 -> /home/ref_genew/Brachypodium_Bd21_v3.1/Bowtie2Index/Bdistachyon314_Bd.1.bt2\n",
      "lrwxrwxrwx 1 feng feng 75 Jun  8 12:15 /home/feng/ref/Brachypodium_Bd21_v3-1/sequence/Bowtie2Index/Bdistachyon314_Bd.2.bt2 -> /home/ref_genew/Brachypodium_Bd21_v3.1/Bowtie2Index/Bdistachyon314_Bd.2.bt2\n",
      "lrwxrwxrwx 1 feng feng 75 Jun  8 12:15 /home/feng/ref/Brachypodium_Bd21_v3-1/sequence/Bowtie2Index/Bdistachyon314_Bd.3.bt2 -> /home/ref_genew/Brachypodium_Bd21_v3.1/Bowtie2Index/Bdistachyon314_Bd.3.bt2\n",
      "lrwxrwxrwx 1 feng feng 75 Jun  8 12:15 /home/feng/ref/Brachypodium_Bd21_v3-1/sequence/Bowtie2Index/Bdistachyon314_Bd.4.bt2 -> /home/ref_genew/Brachypodium_Bd21_v3.1/Bowtie2Index/Bdistachyon314_Bd.4.bt2\n",
      "lrwxrwxrwx 1 feng feng 79 Jun  8 12:15 /home/feng/ref/Brachypodium_Bd21_v3-1/sequence/Bowtie2Index/Bdistachyon314_Bd.rev.1.bt2 -> /home/ref_genew/Brachypodium_Bd21_v3.1/Bowtie2Index/Bdistachyon314_Bd.rev.1.bt2\n",
      "lrwxrwxrwx 1 feng feng 79 Jun  8 12:15 /home/feng/ref/Brachypodium_Bd21_v3-1/sequence/Bowtie2Index/Bdistachyon314_Bd.rev.2.bt2 -> /home/ref_genew/Brachypodium_Bd21_v3.1/Bowtie2Index/Bdistachyon314_Bd.rev.2.bt2\n",
      "271163419\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    ". /home/feng/envs/ref/bin/activate\n",
    ". config_default.sh\n",
    "size2sum $GSIZE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 289,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting /home/feng/repos/BrachyPhoton/pipeline_rnaseq/pipeline_callPeak.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile ~/repos/BrachyPhoton/pipeline_rnaseq/pipeline_callPeak.sh\n",
    "#!/bin/bash\n",
    "# %%bash\n",
    "main(){\n",
    "    set -e\n",
    "    local SELF=`readlink -f ${BASH_SOURCE[0]}`        \n",
    "    local SELFALI=`basename ${SELF%.*}`\n",
    "    \n",
    "    show_help(){\n",
    "    echo \" No help for you yet :X\n",
    "Usage:\n",
    "    pipeline_callPeak.sh [OPTIONS] [-c control.bam] <IN.bam> \n",
    "Options:\n",
    "    -d dry-run\n",
    "    -r [RADIUS]  peaks within this radius will be reported\n",
    "    -q [Q_cutoff] peaks with Q-value lower than the cutoff will be kept\n",
    "    \n",
    "Required environment variables:\n",
    "    GSIZE: *.sizes, size of each chromosome\n",
    "    GFF: GFF annotation \n",
    "    \" \n",
    "        \n",
    "    checkVars GSIZE GFF\n",
    "        \n",
    "    }\n",
    "    if [[ $# -eq 0 ]] ; then\n",
    "        show_help\n",
    "        exit 0\n",
    "    fi\n",
    "    \n",
    "    ###############\n",
    "    #### ==== Parsing argument\n",
    "    local CTRL=\"\"\n",
    "    local GSUM=2.7E9\n",
    "    local PAD=1000\n",
    "    local QLIM=5E-2\n",
    "    local TARGET=\"\"\n",
    "         \n",
    "    OPTIND=1         # Reset in case getopts has been used previously in the shell.\n",
    "    while getopts \"h?c:j:g:r:q:dt\" opt; do\n",
    "        case \"$opt\" in\n",
    "        h|\\?)\n",
    "            show_help\n",
    "            exit 0\n",
    "            ;;\n",
    "        c)  CTRL=$OPTARG\n",
    "            ;;\n",
    "        j)  NCORE=$OPTARG\n",
    "            ;;\n",
    "        g)  GSUM=$OPTARG\n",
    "            ;;\n",
    "        r)  PAD=$OPTARG\n",
    "            ;;\n",
    "        q)  QLIM=$OPTARG\n",
    "            ;;\n",
    "        d)  local DRY=1\n",
    "            ;;\n",
    "        t)  TARGET=1\n",
    "            ;;\n",
    "        esac\n",
    "    done\n",
    "    shift $((OPTIND-1))\n",
    "    [ \"${1:-}\" = \"--\" ] && shift\n",
    "    #### ---- Parsing argument\n",
    "    ###############\n",
    "    echo \"[Positional]: $@\"\n",
    "    ## dbg\n",
    "#     checkVars QLIM \n",
    "    IN=$1 ; \n",
    "    if [[ -d $IN ]]; then\n",
    "        IN=`find $IN -type f -wholename \"*.bam\" | head -n1 `\n",
    "    else\n",
    "         echo [NOT A DIR?]\n",
    "    fi         \n",
    "    IN=`readlink -f $IN`\n",
    "    ALI=`basename ${IN%.*}`     \n",
    "    echo [ALI=$ALI]\n",
    "    checkVars IN   \n",
    "         \n",
    "    #### Reference annotation\n",
    "    grep gene $GFF | bedtools sort  >tmp.ref\n",
    "\n",
    "    {\n",
    "        echo \"#### Step 1: MACS2 callpeak \"\n",
    "        PROG=\"macs2 callpeak\"\n",
    "        PROGALI=${PROG// /_}\n",
    "        OPT=\"-q $QLIM -g $GSUM --keep-dup 1\"\n",
    "\n",
    "        ### Use a control if specified\n",
    "        if [[ ! -z \"$CTRL\" ]]; then\n",
    "            OPT=\"$OPT -c $CTRL\"\n",
    "            ALI=\"${ALI}_ctrl:`basename ${CTRL%.*}`\"\n",
    "        else\n",
    "            ALI=${ALI}_ctrl:none\n",
    "        fi\n",
    "\n",
    "        CMD=\"$PROG $OPT -t $IN -n $ALI &>$ALI.$PROGALI.log\"\n",
    "        echo $CMD ; [[ ! -z $DRY ]] || eval $CMD    \n",
    "\n",
    "        echo \"#### Step 2:  Inner join annotation (GFF) to called peaks\"\n",
    "        INBED=${ALI}_peaks.narrowPeak       \n",
    "        ### tabCount() used since The output of 'bedtools closest' is not dircetly accepted by bedtools ;\n",
    "        CMD=\"\n",
    "        bedtools slop -i tmp.ref -g $GSIZE -b $PAD >tmp.ref.s${PAD} ;\n",
    "        bedtools closest -a tmp.ref.s${PAD} -b $INBED >tmp.out   ;\n",
    "\n",
    "        tabCut tmp.out -9  | bedtools slop -g $GSIZE -b -$PAD -i - >tmp.out.-9 ; \n",
    "        tabCut tmp.out 10- | paste -d$'\\t' tmp.out.-9 - >tmp.bed ;\n",
    "        ln -f tmp.bed ${ALI}.bed ;\n",
    "        \"\n",
    "        echo $CMD; [[ ! -z $DRY ]] || eval $CMD   \n",
    "\n",
    "        echo \"#### Step 3: Select the peak with highest\"\n",
    "        CMD=\"uniqMaxCol.py ${ALI}.bed ${ALI}_uniq.bed  2>${ALI}.uniq.log;\" \n",
    "        echo $CMD; [[ ! -z $DRY ]] || eval $CMD   \n",
    "\n",
    "\n",
    "        echo \"#### Step 4: Output to dir\"\n",
    "        {\n",
    "            OUTDIR=\"script:${SELFALI}_sample:${ALI}\"\n",
    "            CMD=\"mkdir -p $OUTDIR ; \n",
    "            cp *.log *.bed -t $OUTDIR\"\n",
    "            echo $CMD ; [[ ! -z $DRY ]] || eval $CMD\n",
    "            \n",
    "            REMDIR=`dirname $IN`                   \n",
    "            CMD=\"cp -R `readlink -f $OUTDIR` -t $REMDIR\"  \n",
    "            echo \"$CMD\" | tee push.sh\n",
    "            if [[ ! -z \"$TARGET\" ]]; then\n",
    "                [[ ! -z $DRY ]] || . push.sh\n",
    "            fi        \n",
    "        }\n",
    "        echo \"[SUCC][DRY=$DRY] MACS2 pipeline finished for bam input:$IN\"\n",
    "    }\n",
    "    set +e\n",
    "}\n",
    "\n",
    "main \"$@\"\n",
    "         "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 290,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Script is being sourced\n",
      "$BASH_SOURCE /home/feng/envs/ref/bin/activate\n",
      "/home/feng/envs/ref/bin\n",
      "[SOURCE]ing [UTILITIES] from /home/feng/envs/ref/bin/util.sh\n",
      "Adding /home/feng/envs/ref to PATH\n",
      "Script is being sourced\n",
      "$BASH_SOURCE /home/feng/envs/ref/bin/activate\n",
      "/home/feng/envs/ref/bin\n",
      "[SOURCE]ing [UTILITIES] from /home/feng/envs/ref/bin/util.sh\n",
      "Adding /home/feng/envs/ref to PATH\n",
      "A=/home/feng/ref/Brachypodium_Bd21_v3-1/sequence/Bowtie2Index/Bdistachyon314_Bd.1.bt2\n",
      "[Test] GTF=/home/feng/ref/Brachypodium_Bd21_v3-1/annotation/Bdistachyon_314_v3.1.gene_exons.gtf\n",
      "[Test] GFF=/home/feng/ref/Brachypodium_Bd21_v3-1/annotation/Bdistachyon_314_v3.1.gene_exons.gff3\n",
      "[Test] GSIZE=/home/feng/ref/Brachypodium_Bd21_v3-1/genome.sizes\n",
      "[Test] FA_ADAPTER=/home/Program_NGS_sl-pw-srv01/Trimmomatic-0.32/adapters/TruSeq3-SE.fa\n",
      "[Test] REF=/home/feng/ref/Brachypodium_Bd21_v3-1\n",
      "[Test] IDX_BOWTIE2=/home/feng/ref/Brachypodium_Bd21_v3-1/sequence/Bowtie2Index/Bdistachyon314_Bd\n",
      "lrwxrwxrwx 1 feng feng 75 Jun  8 12:15 /home/feng/ref/Brachypodium_Bd21_v3-1/sequence/Bowtie2Index/Bdistachyon314_Bd.1.bt2 -> /home/ref_genew/Brachypodium_Bd21_v3.1/Bowtie2Index/Bdistachyon314_Bd.1.bt2\n",
      "lrwxrwxrwx 1 feng feng 75 Jun  8 12:15 /home/feng/ref/Brachypodium_Bd21_v3-1/sequence/Bowtie2Index/Bdistachyon314_Bd.2.bt2 -> /home/ref_genew/Brachypodium_Bd21_v3.1/Bowtie2Index/Bdistachyon314_Bd.2.bt2\n",
      "lrwxrwxrwx 1 feng feng 75 Jun  8 12:15 /home/feng/ref/Brachypodium_Bd21_v3-1/sequence/Bowtie2Index/Bdistachyon314_Bd.3.bt2 -> /home/ref_genew/Brachypodium_Bd21_v3.1/Bowtie2Index/Bdistachyon314_Bd.3.bt2\n",
      "lrwxrwxrwx 1 feng feng 75 Jun  8 12:15 /home/feng/ref/Brachypodium_Bd21_v3-1/sequence/Bowtie2Index/Bdistachyon314_Bd.4.bt2 -> /home/ref_genew/Brachypodium_Bd21_v3.1/Bowtie2Index/Bdistachyon314_Bd.4.bt2\n",
      "lrwxrwxrwx 1 feng feng 79 Jun  8 12:15 /home/feng/ref/Brachypodium_Bd21_v3-1/sequence/Bowtie2Index/Bdistachyon314_Bd.rev.1.bt2 -> /home/ref_genew/Brachypodium_Bd21_v3.1/Bowtie2Index/Bdistachyon314_Bd.rev.1.bt2\n",
      "lrwxrwxrwx 1 feng feng 79 Jun  8 12:15 /home/feng/ref/Brachypodium_Bd21_v3-1/sequence/Bowtie2Index/Bdistachyon314_Bd.rev.2.bt2 -> /home/ref_genew/Brachypodium_Bd21_v3.1/Bowtie2Index/Bdistachyon314_Bd.rev.2.bt2\n",
      " No help for you yet :X\n",
      "Usage:\n",
      "    pipeline_callPeak.sh [OPTIONS] [-c control.bam] <IN.bam> \n",
      "Options:\n",
      "    -d dry-run\n",
      "    -r [RADIUS]  peaks within this radius will be reported\n",
      "    -q [Q_cutoff] peaks with Q-value lower than the cutoff will be kept\n",
      "    \n",
      "Required environment variables:\n",
      "    GSIZE: *.sizes, size of each chromosome\n",
      "    GFF: GFF annotation \n",
      "    \n",
      "[Test] GSIZE=/home/feng/ref/Brachypodium_Bd21_v3-1/genome.sizes\n",
      "[Test] GFF=/home/feng/ref/Brachypodium_Bd21_v3-1/annotation/Bdistachyon_314_v3.1.gene_exons.gff3\n",
      "[Positional]: ../test.bam\n",
      "[NOT A DIR?]\n",
      "[ALI=test]\n",
      "[Test] IN=/home/feng/envs/callPeak/test.bam\n",
      "#### Step 1: MACS2 callpeak \n",
      "macs2 callpeak -q 5E-4 -g 2.7E9 --keep-dup 1 -t /home/feng/envs/callPeak/test.bam -n test_ctrl:none &>test_ctrl:none.macs2_callpeak.log\n",
      "#### Step 2:  Inner join annotation (GFF) to called peaks\n",
      "bedtools slop -i tmp.ref -g /home/feng/ref/Brachypodium_Bd21_v3-1/genome.sizes -b 1000 >tmp.ref.s1000 ; bedtools closest -a tmp.ref.s1000 -b test_ctrl:none_peaks.narrowPeak >tmp.out ; tabCut tmp.out -9 | bedtools slop -g /home/feng/ref/Brachypodium_Bd21_v3-1/genome.sizes -b -1000 -i - >tmp.out.-9 ; tabCut tmp.out 10- | paste -d$'\\t' tmp.out.-9 - >tmp.bed ; ln -f tmp.bed test_ctrl:none.bed ;\n",
      "#### Step 3: Select the peak with highest\n",
      "uniqMaxCol.py test_ctrl:none.bed test_ctrl:none_uniq.bed 2>test_ctrl:none.uniq.log;\n",
      "#### Step 4: Output to dir\n",
      "mkdir -p script:pipeline_callPeak_sample:test_ctrl:none ; cp test_ctrl:none.macs2_callpeak.log test_ctrl:none.uniq.log test_ctrl:none.bed test_ctrl:none_summits.bed test_ctrl:none_uniq.bed tmp.bed -t script:pipeline_callPeak_sample:test_ctrl:none\n",
      "cp -R /home/feng/envs/callPeak/test/script:pipeline_callPeak_sample:test_ctrl:none -t /home/feng/envs/callPeak\n",
      "[SUCC][DRY=] MACS2 pipeline finished for bam input:/home/feng/envs/callPeak/test.bam\n",
      "-rw-rw-r-- 1 feng feng 3.8M Jun 11 10:34 /home/feng/envs/callPeak/test/tmp.ref\n",
      "-rw-rw-r-- 1 feng feng  77K Jun 11 10:34 /home/feng/envs/callPeak/test/test_ctrl:none_model.r\n",
      "-rw-rw-r-- 1 feng feng  30K Jun 11 10:34 /home/feng/envs/callPeak/test/test_ctrl:none_peaks.xls\n",
      "-rw-rw-r-- 1 feng feng  27K Jun 11 10:34 /home/feng/envs/callPeak/test/test_ctrl:none_peaks.narrowPeak\n",
      "-rw-rw-r-- 1 feng feng  18K Jun 11 10:34 /home/feng/envs/callPeak/test/test_ctrl:none_summits.bed\n",
      "-rw-rw-r-- 1 feng feng 2.9K Jun 11 10:34 /home/feng/envs/callPeak/test/test_ctrl:none.macs2_callpeak.log\n",
      "-rw-rw-r-- 1 feng feng 3.8M Jun 11 10:34 /home/feng/envs/callPeak/test/tmp.ref.s1000\n",
      "-rw-rw-r-- 1 feng feng 6.5M Jun 11 10:34 /home/feng/envs/callPeak/test/tmp.out\n",
      "-rw-rw-r-- 1 feng feng 3.8M Jun 11 10:34 /home/feng/envs/callPeak/test/tmp.out.-9\n",
      "-rw-rw-r-- 2 feng feng 6.5M Jun 11 10:34 /home/feng/envs/callPeak/test/tmp.bed\n",
      "-rw-rw-r-- 2 feng feng 6.5M Jun 11 10:34 /home/feng/envs/callPeak/test/test_ctrl:none.bed\n",
      "-rw-rw-r-- 1 feng feng    0 Jun 11 10:34 /home/feng/envs/callPeak/test/test_ctrl:none.uniq.log\n",
      "-rw-rw-r-- 1 feng feng 6.5M Jun 11 10:34 /home/feng/envs/callPeak/test/test_ctrl:none_uniq.bed\n",
      "-rw-rw-r-- 1 feng feng  111 Jun 11 10:34 /home/feng/envs/callPeak/test/push.sh\n",
      "\n",
      "/home/feng/envs/callPeak/test/script:pipeline_callPeak_sample:test_ctrl:none:\n",
      "total 20M\n",
      "-rw-rw-r-- 1 feng feng    0 Jun 11 10:34 test_ctrl:none.uniq.log\n",
      "-rw-rw-r-- 1 feng feng 2.9K Jun 11 10:34 test_ctrl:none.macs2_callpeak.log\n",
      "-rw-rw-r-- 1 feng feng  18K Jun 11 10:34 test_ctrl:none_summits.bed\n",
      "-rw-rw-r-- 1 feng feng 6.5M Jun 11 10:34 test_ctrl:none.bed\n",
      "-rw-rw-r-- 1 feng feng 6.5M Jun 11 10:34 test_ctrl:none_uniq.bed\n",
      "-rw-rw-r-- 1 feng feng 6.5M Jun 11 10:34 tmp.bed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cp: omitting directory '/home/feng/repos/BrachyPhoton/pipeline_rnaseq/config'\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    ". /home/feng/envs/ref/bin/activate ; envPull\n",
    ". /home/feng/envs/ref/bin/activate ; . config_default.sh \n",
    "\n",
    "# ls -lhtr .\n",
    "rm -rf test/\n",
    "PROG=pipeline_callPeak.sh\n",
    "$PROG\n",
    "# $PROG -d -c hi.bam test.bam\n",
    "# $PROG -d $PWD\n",
    "# exit 0 \n",
    "mkdir -p test/ ; cd test/\n",
    "\n",
    "# $PROG -q 5E-2 -c ../test.bam ../test.bam\n",
    "$PROG -q 5E-4 ../test.bam \n",
    "ls -lhtr $PWD/*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "    5th: integer score for display calculated as int(-10*log10qvalue). Please note that currently this value might be out of the [0-1000] range defined in UCSC Encode narrowPeak format<https://genome.ucsc.edu/FAQ/FAQformat.html#format12>\n",
    "    7th: fold-change\n",
    "    8th: -log10pvalue\n",
    "    9th: -log10qvalue\n",
    "    10th: relative summit position to peak start\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Script is being sourced\n",
      "$BASH_SOURCE /home/feng/envs/ref/bin/activate\n",
      "/home/feng/envs/ref/bin\n",
      "[SOURCE]ing [UTILITIES] from /home/feng/envs/ref/bin/util.sh\n",
      "Adding /home/feng/envs/ref to PATH\n",
      "A=/home/feng/ref/Brachypodium_Bd21_v3-1/sequence/Bowtie2Index/Bdistachyon314_Bd.1.bt2\n",
      "[Test] GTF=/home/feng/ref/Brachypodium_Bd21_v3-1/annotation/Bdistachyon_314_v3.1.gene_exons.gtf\n",
      "[Test] GFF=/home/feng/ref/Brachypodium_Bd21_v3-1/annotation/Bdistachyon_314_v3.1.gene_exons.gff3\n",
      "[Test] GSIZE=/home/feng/ref/Brachypodium_Bd21_v3-1/genome.sizes\n",
      "[Test] FA_ADAPTER=/home/Program_NGS_sl-pw-srv01/Trimmomatic-0.32/adapters/TruSeq3-SE.fa\n",
      "[Test] REF=/home/feng/ref/Brachypodium_Bd21_v3-1\n",
      "[Test] IDX_BOWTIE2=/home/feng/ref/Brachypodium_Bd21_v3-1/sequence/Bowtie2Index/Bdistachyon314_Bd\n",
      "lrwxrwxrwx 1 feng feng 75 Jun  8 12:15 /home/feng/ref/Brachypodium_Bd21_v3-1/sequence/Bowtie2Index/Bdistachyon314_Bd.1.bt2 -> /home/ref_genew/Brachypodium_Bd21_v3.1/Bowtie2Index/Bdistachyon314_Bd.1.bt2\n",
      "lrwxrwxrwx 1 feng feng 75 Jun  8 12:15 /home/feng/ref/Brachypodium_Bd21_v3-1/sequence/Bowtie2Index/Bdistachyon314_Bd.2.bt2 -> /home/ref_genew/Brachypodium_Bd21_v3.1/Bowtie2Index/Bdistachyon314_Bd.2.bt2\n",
      "lrwxrwxrwx 1 feng feng 75 Jun  8 12:15 /home/feng/ref/Brachypodium_Bd21_v3-1/sequence/Bowtie2Index/Bdistachyon314_Bd.3.bt2 -> /home/ref_genew/Brachypodium_Bd21_v3.1/Bowtie2Index/Bdistachyon314_Bd.3.bt2\n",
      "lrwxrwxrwx 1 feng feng 75 Jun  8 12:15 /home/feng/ref/Brachypodium_Bd21_v3-1/sequence/Bowtie2Index/Bdistachyon314_Bd.4.bt2 -> /home/ref_genew/Brachypodium_Bd21_v3.1/Bowtie2Index/Bdistachyon314_Bd.4.bt2\n",
      "lrwxrwxrwx 1 feng feng 79 Jun  8 12:15 /home/feng/ref/Brachypodium_Bd21_v3-1/sequence/Bowtie2Index/Bdistachyon314_Bd.rev.1.bt2 -> /home/ref_genew/Brachypodium_Bd21_v3.1/Bowtie2Index/Bdistachyon314_Bd.rev.1.bt2\n",
      "lrwxrwxrwx 1 feng feng 79 Jun  8 12:15 /home/feng/ref/Brachypodium_Bd21_v3-1/sequence/Bowtie2Index/Bdistachyon314_Bd.rev.2.bt2 -> /home/ref_genew/Brachypodium_Bd21_v3.1/Bowtie2Index/Bdistachyon314_Bd.rev.2.bt2\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    ". /home/feng/envs/ref/bin/activate ; . config_default.sh \n",
    "# GREF=/home/feng/ref/Brachypodium_Bd21_v3-1/annotation/gene_only.gff\n",
    "# echo $GFF\n",
    "\n",
    "IN=NA_peaks.narrowPeak\n",
    "# echo $IN\n",
    "\n",
    "ALI=`bname ${IN}`\n",
    "#### Reference annotation\n",
    "GREF=/home/feng/ref/Brachypodium_Bd21_v3-1/annotation/Bdistachyon_314_v3.1.gene.gff3\n",
    "grep gene $GREF | bedtools sort  >tmp.ref\n",
    "PAD=${PAD:-1000}\n",
    "\n",
    "# CMD=\"\n",
    "bedtools slop -i tmp.ref -g $GSIZE -b $PAD >tmp.ref.s${PAD}\n",
    "bedtools closest -a tmp.ref.s${PAD} -b $IN >tmp.out  \n",
    "\n",
    "#### The output of 'bedtools closest' is not dircetly accepted by bedtools\n",
    "tabCut tmp.out -9  | bedtools slop -i - -g $GSIZE -b -${PAD} >tmp.out.-9\n",
    "tabCut tmp.out 10- | paste -d$'\\t' tmp.out.-9 - >tmp.bed\n",
    "ln -f tmp.bed ${ALI}.bed\n",
    "\n",
    "uniqMaxCol.py ${ALI}.bed ${ALI}_uniq.bed \n",
    "# \"\n",
    "# eval $CMD\n",
    "\n",
    "mkdir -p output\n",
    "cp -f *.log *.bed -t output\n",
    "#### Peaks are associated to each gene\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-rw-r-- 1 feng feng  14K Jun  8 18:13 HELP\n",
      "-rw-rw-r-- 2 feng feng 107M Jun  8 18:17 test.bam\n",
      "-rw-rw-r-- 2 feng feng 107M Jun  8 18:17 ctrl.bam\n",
      "-rw-rw-r-- 1 feng feng  77K Jun  8 18:35 NA_model.r\n",
      "-rw-rw-r-- 1 feng feng  40K Jun  8 18:35 NA_peaks.xls\n",
      "-rw-rw-r-- 1 feng feng  34K Jun  8 18:35 NA_peaks.narrowPeak\n",
      "-rw-rw-r-- 1 feng feng  21K Jun  8 18:35 NA_summits.bed\n",
      "-rw-rw-r-- 1 feng feng  77K Jun  8 18:59 test_model.r\n",
      "-rw-rw-r-- 1 feng feng  41K Jun  8 18:59 test_peaks.xls\n",
      "-rw-rw-r-- 1 feng feng  35K Jun  8 18:59 test_peaks.narrowPeak\n",
      "-rw-rw-r-- 1 feng feng  22K Jun  8 18:59 test_summits.bed\n",
      "-rw-rw-r-- 1 feng feng 2.8K Jun  8 18:59 test.macs2_callpeak.log\n",
      "-rw-rw-r-- 1 feng feng 7.2K Jun  8 19:36 tmp\n",
      "-rw-rw-r-- 1 feng feng 3.8M Jun  8 19:40 tmp_s1000.ref\n",
      "-rw-rw-r-- 1 feng feng 3.8M Jun  8 19:59 tmp.cut\n",
      "-rw-rw-r-- 1 feng feng 6.1M Jun  9 18:36 tmp_dedup.bed\n",
      "-rwxrwxr-x 1 feng feng 7.5K Jun  9 18:40 uniqMaxCol.py\n",
      "-rw-rw-r-- 1 feng feng 3.8M Jun  9 18:43 tmp.ref\n",
      "-rw-rw-r-- 1 feng feng 3.8M Jun  9 18:43 tmp.ref.s1000\n",
      "-rw-rw-r-- 1 feng feng 6.1M Jun  9 18:43 tmp.out\n",
      "-rw-rw-r-- 1 feng feng 3.8M Jun  9 18:43 tmp.out.-9\n",
      "-rw-rw-r-- 2 feng feng 6.1M Jun  9 18:43 tmp.bed\n",
      "-rw-rw-r-- 2 feng feng 6.1M Jun  9 18:43 NA_peaks.bed\n",
      "-rw-rw-r-- 1 feng feng 6.1M Jun  9 18:43 NA_peaks_uniq.bed\n",
      "-rw-rw-r-- 1 feng feng  80K Jun  9 18:43 uniqMaxCol.ipynb\n",
      "\n",
      "raw:\n",
      "total 441M\n",
      "-rwxrwxr-x 1 feng feng 204K Jun  8 18:05 Bd-ELF3OX-SD-ZT16-_S2.bai\n",
      "-rwxrwxr-x 1 feng feng 261M Jun  8 18:05 Bd-ELF3OX-SD-ZT16-_S2.bam\n",
      "-rwxrwxr-x 1 feng feng 120M Jun  8 18:05 Bd-ELF3OX-SD-ZT16-_S2.bdg\n",
      "-rwxrwxr-x 1 feng feng 1.1M Jun  8 18:05 Bd-ELF3OX-SD-ZT16-_S2.bowtie2.log\n",
      "-rwxrwxr-x 1 feng feng 1.1K Jun  8 18:05 Bd-ELF3OX-SD-ZT16-_S2.dupstat.log\n",
      "-rwxrwxr-x 1 feng feng    0 Jun  8 18:05 Bd-ELF3OX-SD-ZT16-_S2.cleanup.log\n",
      "-rwxrwxr-x 1 feng feng  32M Jun  8 18:05 Bd-ELF3OX-SD-ZT16-_S2.bw\n",
      "-rwxrwxr-x 1 feng feng 3.9K Jun  8 18:05 Bd-ELF3OX-SD-ZT16-_S2.pipeline_rnaseq.log\n",
      "-rwxrwxr-x 1 feng feng 3.8K Jun  8 18:05 Bd-ELF3OX-SD-ZT16-_S2.picard.log\n",
      "-rwxrwxr-x 1 feng feng  387 Jun  8 18:05 Bd-ELF3OX-SD-ZT16-_S2.flagstat.log\n",
      "-rwxrwxr-x 1 feng feng 355K Jun  8 18:05 Bd-ELF3OX-SD-ZT16-_S2_R1_raw.fastq.100k.tmp_fastqc.html\n",
      "-rwxrwxr-x 1 feng feng 336K Jun  8 18:05 Bd-ELF3OX-SD-ZT16-_S2_R1_raw_pass.fastq.100k.tmp_fastqc.html\n",
      "-rwxrwxr-x 1 feng feng  401 Jun  8 18:05 Bd-ELF3OX-SD-ZT16-_S2.time\n",
      "-rwxrwxr-x 1 feng feng 347K Jun  8 18:05 Bd-ELF3OX-SD-ZT16-_S2_R1_raw_pass_fastqc.html\n",
      "-rwxrwxr-x 1 feng feng  26M Jun  8 18:05 Bd-ELF3OX-SD-ZT16-_S2_RPGC.bw\n",
      "\n",
      "output:\n",
      "total 25M\n",
      "-rw-rw-r-- 1 feng feng 2.8K Jun  9 18:43 test.macs2_callpeak.log\n",
      "-rw-rw-r-- 1 feng feng 6.1M Jun  9 18:43 NA_peaks.bed\n",
      "-rw-rw-r-- 1 feng feng  22K Jun  9 18:43 test_summits.bed\n",
      "-rw-rw-r-- 1 feng feng  21K Jun  9 18:43 NA_summits.bed\n",
      "-rw-rw-r-- 1 feng feng 6.1M Jun  9 18:43 NA_peaks_uniq.bed\n",
      "-rw-rw-r-- 1 feng feng 6.1M Jun  9 18:43 tmp.bed\n",
      "-rw-rw-r-- 1 feng feng 6.1M Jun  9 18:43 tmp_dedup.bed\n",
      "   34374 NA_peaks.bed\n",
      "   34310 NA_peaks_uniq.bed\n",
      "     509 NA_summits.bed\n",
      "     509 test_summits.bed\n",
      "   34374 tmp.bed\n",
      "   34310 tmp_dedup.bed\n",
      "  138386 total\n"
     ]
    }
   ],
   "source": [
    "! ls -lhtr *\n",
    "! wc -l *.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 308,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['/home/feng/.local/lib/python2.7/site-packages/ipykernel_launcher.py', '-f', '/run/user/1039/jupyter/kernel-ea7b5ce5-4c19-44b3-89c6-c677db1f0043.json']\n"
     ]
    }
   ],
   "source": [
    "df = uniqMaxCol(fname='tmp.bed',)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bd1\tphytozomev10\tgene\t10581\t11638\t.\t+\t.\tID=Bradi1g00200.v3.1;Name=Bradi1g00200;ancestorIdentifier=Bradi1g00200.v2.1\tBd1\t9738\t10049\tNA_peak_2\t45\t.\t5.63936\t7.79234\t4.56778\t20\n",
      "Bd1\tphytozomev10\tgene\t18453\t21742\t.\t+\t.\tID=Bradi1g00210.v3.1;Name=Bradi1g00210;ancestorIdentifier=Bradi1g00210.v2.1\tBd1\t9738\t10049\tNA_peak_2\t45\t.\t5.63936\t7.79234\t4.56778\t20\n",
      "Bd1\tphytozomev10\tgene\t28710\t31462\t.\t-\t.\tID=Bradi1g00213.v3.1;Name=Bradi1g00213;ancestorIdentifier=Bradi1g00213.v2.1\tBd1\t32587\t33418\tNA_peak_3\t138\t.\t7.37483\t17.67282\t13.80929\t292\n",
      "Bd1\tphytozomev10\tgene\t47522\t49251\t.\t+\t.\tID=Bradi1g00215.v3.1;Name=Bradi1g00215\tBd1\t33530\t33883\tNA_peak_4\t59\t.\t4.95704\t9.32209\t5.98403\t197\n",
      "Bd1\tphytozomev10\tgene\t52489\t52869\t.\t+\t.\tID=Bradi1g00217.v3.1;Name=Bradi1g00217\tBd1\t33530\t33883\tNA_peak_4\t59\t.\t4.95704\t9.32209\t5.98403\t197\n",
      "Bd1\tphytozomev10\tgene\t54094\t56379\t.\t+\t.\tID=Bradi1g00219.v3.1;Name=Bradi1g00219;ancestorIdentifier=Bradi1g00219.v2.1\tBd1\t33530\t33883\tNA_peak_4\t59\t.\t4.95704\t9.32209\t5.98403\t197\n",
      "Bd1\tphytozomev10\tgene\t56819\t57537\t.\t-\t.\tID=Bradi1g00222.v3.1;Name=Bradi1g00222;ancestorIdentifier=Bradi1g00222.v2.1\tBd1\t33530\t33883\tNA_peak_4\t59\t.\t4.95704\t9.32209\t5.98403\t197\n",
      "Bd1\tphytozomev10\tgene\t59795\t70558\t.\t+\t.\tID=Bradi1g00234.v3.1;Name=Bradi1g00234\tBd1\t33530\t33883\tNA_peak_4\t59\t.\t4.95704\t9.32209\t5.98403\t197\n",
      "Bd1\tphytozomev10\tgene\t104368\t106734\t.\t-\t.\tID=Bradi1g00247.v3.1;Name=Bradi1g00247;ancestorIdentifier=Bradi1g00247.v2.1\tBd1\t33530\t33883\tNA_peak_4\t59\t.\t4.95704\t9.32209\t5.98403\t197\n",
      "Bd1\tphytozomev10\tgene\t108527\t109357\t.\t+\t.\tID=Bradi1g00253.v3.1;Name=Bradi1g00253;ancestorIdentifier=Bradi1g00253.v2.1\tBd1\t33530\t33883\tNA_peak_4\t59\t.\t4.95704\t9.32209\t5.98403\t197\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "***** ERROR: Unrecognized parameter: tmp.bed *****\n",
      "\n",
      "Tool:    bedtools merge (aka mergeBed)\n",
      "Version: v2.26.0\n",
      "Summary: Merges overlapping BED/GFF/VCF entries into a single interval.\n",
      "\n",
      "Usage:   bedtools merge [OPTIONS] -i <bed/gff/vcf>\n",
      "\n",
      "Options: \n",
      "\t-s\tForce strandedness.  That is, only merge features\n",
      "\t\tthat are on the same strand.\n",
      "\t\t- By default, merging is done without respect to strand.\n",
      "\n",
      "\t-S\tForce merge for one specific strand only.\n",
      "\t\tFollow with + or - to force merge from only\n",
      "\t\tthe forward or reverse strand, respectively.\n",
      "\t\t- By default, merging is done without respect to strand.\n",
      "\n",
      "\t-d\tMaximum distance between features allowed for features\n",
      "\t\tto be merged.\n",
      "\t\t- Def. 0. That is, overlapping & book-ended features are merged.\n",
      "\t\t- (INTEGER)\n",
      "\t\t- Note: negative values enforce the number of b.p. required for overlap.\n",
      "\n",
      "\t-c\tSpecify columns from the B file to map onto intervals in A.\n",
      "\t\tDefault: 5.\n",
      "\t\tMultiple columns can be specified in a comma-delimited list.\n",
      "\n",
      "\t-o\tSpecify the operation that should be applied to -c.\n",
      "\t\tValid operations:\n",
      "\t\t    sum, min, max, absmin, absmax,\n",
      "\t\t    mean, median, mode, antimode\n",
      "\t\t    stdev, sstdev\n",
      "\t\t    collapse (i.e., print a delimited list (duplicates allowed)), \n",
      "\t\t    distinct (i.e., print a delimited list (NO duplicates allowed)), \n",
      "\t\t    distinct_sort_num (as distinct, sorted numerically, ascending),\n",
      "\t\t    distinct_sort_num_desc (as distinct, sorted numerically, desscending),\n",
      "\t\t    distinct_only (delimited list of only unique values),\n",
      "\t\t    count\n",
      "\t\t    count_distinct (i.e., a count of the unique values in the column), \n",
      "\t\t    first (i.e., just the first value in the column), \n",
      "\t\t    last (i.e., just the last value in the column), \n",
      "\t\tDefault: sum\n",
      "\t\tMultiple operations can be specified in a comma-delimited list.\n",
      "\n",
      "\t\tIf there is only column, but multiple operations, all operations will be\n",
      "\t\tapplied on that column. Likewise, if there is only one operation, but\n",
      "\t\tmultiple columns, that operation will be applied to all columns.\n",
      "\t\tOtherwise, the number of columns must match the the number of operations,\n",
      "\t\tand will be applied in respective order.\n",
      "\t\tE.g., \"-c 5,4,6 -o sum,mean,count\" will give the sum of column 5,\n",
      "\t\tthe mean of column 4, and the count of column 6.\n",
      "\t\tThe order of output columns will match the ordering given in the command.\n",
      "\n",
      "\n",
      "\t-delim\tSpecify a custom delimiter for the collapse operations.\n",
      "\t\t- Example: -delim \"|\"\n",
      "\t\t- Default: \",\".\n",
      "\n",
      "\t-prec\tSets the decimal precision for output (Default: 5)\n",
      "\n",
      "\t-bed\tIf using BAM input, write output as BED.\n",
      "\n",
      "\t-header\tPrint the header from the A file prior to results.\n",
      "\n",
      "\t-nobuf\tDisable buffered output. Using this option will cause each line\n",
      "\t\tof output to be printed as it is generated, rather than saved\n",
      "\t\tin a buffer. This will make printing large output files \n",
      "\t\tnoticeably slower, but can be useful in conjunction with\n",
      "\t\tother software tools and scripts that need to process one\n",
      "\t\tline of bedtools output at a time.\n",
      "\n",
      "\t-iobuf\tSpecify amount of memory to use for input buffer.\n",
      "\t\tTakes an integer argument. Optional suffixes K/M/G supported.\n",
      "\t\tNote: currently has no effect with compressed files.\n",
      "\n",
      "Notes: \n",
      "\t(1) The input file (-i) file must be sorted by chrom, then start.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# %%bash \n",
    "# head tmp.`bed\n",
    "# bedtools merge -i tmp.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bd1\tphytozomev10\tgene\t9581\t12638\t.\t+\t.\tID=Bradi1g00200.v3.1;Name=Bradi1g00200;ancestorIdentifier=Bradi1g00200.v2.1\tBd1\t9738\t10049\tNA_peak_2\t45\t.\t5.63936\t7.79234\t4.56778\t20\n",
      "Bd1\tphytozomev10\tgene\t10581\t11638\t.\t+\t.\tID=Bradi1g00200.v3.1;Name=Bradi1g00200;ancestorIdentifier=Bradi1g00200.v2.1\tBd1\t9738\t10049\tNA_peak_2\t45\t.\t5.63936\t7.79234\t4.56778\t20\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "head tmp.out -n1\n",
    "head tmp.bed -n1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-155-cff6cf9e506b>, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-155-cff6cf9e506b>\"\u001b[0;36m, line \u001b[0;32m1\u001b[0m\n\u001b[0;31m    echo =======\u001b[0m\n\u001b[0m            ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "# %%bash\n",
    "# echo =======\n",
    "# head tmp.ref\n",
    "# echo =======\n",
    "# bedtools slop -i tmp.ref.s1000 -g $GSIZE -b -1000 | head \n",
    "# # head tmp.out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##gff-version 3\r\n",
      "##annot-version v3.1\r\n",
      "##species Brachypodium distachyon\r\n",
      "Bd1\tphytozomev10\tgene\t10581\t11638\t.\t+\t.\tID=Bradi1g00200.v3.1;Name=Bradi1g00200;ancestorIdentifier=Bradi1g00200.v2.1\r\n",
      "Bd1\tphytozomev10\tmRNA\t10581\t11638\t.\t+\t.\tID=Bradi1g00200.1.v3.1;Name=Bradi1g00200.1;pacid=32803675;longest=1;ancestorIdentifier=Bradi1g00200.1.v2.1;Parent=Bradi1g00200.v3.1\r\n",
      "Bd1\tphytozomev10\texon\t10581\t10850\t.\t+\t.\tID=Bradi1g00200.1.v3.1.exon.1;Parent=Bradi1g00200.1.v3.1;pacid=32803675\r\n",
      "Bd1\tphytozomev10\tCDS\t10581\t10850\t.\t+\t0\tID=Bradi1g00200.1.v3.1.CDS.1;Parent=Bradi1g00200.1.v3.1;pacid=32803675\r\n",
      "Bd1\tphytozomev10\texon\t11252\t11638\t.\t+\t.\tID=Bradi1g00200.1.v3.1.exon.2;Parent=Bradi1g00200.1.v3.1;pacid=32803675\r\n",
      "Bd1\tphytozomev10\tCDS\t11252\t11638\t.\t+\t0\tID=Bradi1g00200.1.v3.1.CDS.2;Parent=Bradi1g00200.1.v3.1;pacid=32803675\r\n",
      "Bd1\tphytozomev10\tgene\t18453\t21742\t.\t+\t.\tID=Bradi1g00210.v3.1;Name=Bradi1g00210;ancestorIdentifier=Bradi1g00210.v2.1\r\n",
      "Bd1\tphytozomev10\tmRNA\t18453\t21742\t.\t+\t.\tID=Bradi1g00210.3.v3.1;Name=Bradi1g00210.3;pacid=32804311;longest=1;Parent=Bradi1g00210.v3.1\r\n",
      "Bd1\tphytozomev10\texon\t18453\t18628\t.\t+\t.\tID=Bradi1g00210.3.v3.1.exon.1;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\tfive_prime_UTR\t18453\t18498\t.\t+\t.\tID=Bradi1g00210.3.v3.1.five_prime_UTR.1;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\tCDS\t18499\t18628\t.\t+\t0\tID=Bradi1g00210.3.v3.1.CDS.1;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\texon\t18726\t18773\t.\t+\t.\tID=Bradi1g00210.3.v3.1.exon.2;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\tCDS\t18726\t18773\t.\t+\t2\tID=Bradi1g00210.3.v3.1.CDS.2;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\texon\t19132\t19176\t.\t+\t.\tID=Bradi1g00210.3.v3.1.exon.3;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\tCDS\t19132\t19176\t.\t+\t2\tID=Bradi1g00210.3.v3.1.CDS.3;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\texon\t19327\t19446\t.\t+\t.\tID=Bradi1g00210.3.v3.1.exon.4;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\tCDS\t19327\t19446\t.\t+\t2\tID=Bradi1g00210.3.v3.1.CDS.4;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\texon\t19750\t19895\t.\t+\t.\tID=Bradi1g00210.3.v3.1.exon.5;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\tCDS\t19750\t19895\t.\t+\t2\tID=Bradi1g00210.3.v3.1.CDS.5;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\texon\t20030\t20177\t.\t+\t.\tID=Bradi1g00210.3.v3.1.exon.6;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\tCDS\t20030\t20177\t.\t+\t0\tID=Bradi1g00210.3.v3.1.CDS.6;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\texon\t20599\t20682\t.\t+\t.\tID=Bradi1g00210.3.v3.1.exon.7;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\tCDS\t20599\t20682\t.\t+\t2\tID=Bradi1g00210.3.v3.1.CDS.7;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\texon\t20787\t20902\t.\t+\t.\tID=Bradi1g00210.3.v3.1.exon.8;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\tCDS\t20787\t20902\t.\t+\t2\tID=Bradi1g00210.3.v3.1.CDS.8;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\texon\t21165\t21223\t.\t+\t.\tID=Bradi1g00210.3.v3.1.exon.9;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\tCDS\t21165\t21223\t.\t+\t0\tID=Bradi1g00210.3.v3.1.CDS.9;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\texon\t21335\t21425\t.\t+\t.\tID=Bradi1g00210.3.v3.1.exon.10;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\tCDS\t21335\t21425\t.\t+\t1\tID=Bradi1g00210.3.v3.1.CDS.10;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\texon\t21515\t21742\t.\t+\t.\tID=Bradi1g00210.3.v3.1.exon.11;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\tCDS\t21515\t21742\t.\t+\t0\tID=Bradi1g00210.3.v3.1.CDS.11;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\tgene\t28710\t31462\t.\t-\t.\tID=Bradi1g00213.v3.1;Name=Bradi1g00213;ancestorIdentifier=Bradi1g00213.v2.1\r\n",
      "Bd1\tphytozomev10\tmRNA\t28710\t31462\t.\t-\t.\tID=Bradi1g00213.1.v3.1;Name=Bradi1g00213.1;pacid=32803704;longest=1;ancestorIdentifier=Bradi1g00213.1.v2.1;Parent=Bradi1g00213.v3.1\r\n",
      "Bd1\tphytozomev10\texon\t31315\t31462\t.\t-\t.\tID=Bradi1g00213.1.v3.1.exon.1;Parent=Bradi1g00213.1.v3.1;pacid=32803704\r\n",
      "Bd1\tphytozomev10\tCDS\t31315\t31462\t.\t-\t0\tID=Bradi1g00213.1.v3.1.CDS.1;Parent=Bradi1g00213.1.v3.1;pacid=32803704\r\n",
      "Bd1\tphytozomev10\texon\t31064\t31132\t.\t-\t.\tID=Bradi1g00213.1.v3.1.exon.2;Parent=Bradi1g00213.1.v3.1;pacid=32803704\r\n",
      "Bd1\tphytozomev10\tCDS\t31064\t31132\t.\t-\t2\tID=Bradi1g00213.1.v3.1.CDS.2;Parent=Bradi1g00213.1.v3.1;pacid=32803704\r\n",
      "Bd1\tphytozomev10\texon\t30640\t31013\t.\t-\t.\tID=Bradi1g00213.1.v3.1.exon.3;Parent=Bradi1g00213.1.v3.1;pacid=32803704\r\n",
      "Bd1\tphytozomev10\tCDS\t30640\t31013\t.\t-\t2\tID=Bradi1g00213.1.v3.1.CDS.3;Parent=Bradi1g00213.1.v3.1;pacid=32803704\r\n",
      "Bd1\tphytozomev10\texon\t30412\t30570\t.\t-\t.\tID=Bradi1g00213.1.v3.1.exon.4;Parent=Bradi1g00213.1.v3.1;pacid=32803704\r\n",
      "Bd1\tphytozomev10\tCDS\t30412\t30570\t.\t-\t0\tID=Bradi1g00213.1.v3.1.CDS.4;Parent=Bradi1g00213.1.v3.1;pacid=32803704\r\n",
      "Bd1\tphytozomev10\texon\t29955\t30309\t.\t-\t.\tID=Bradi1g00213.1.v3.1.exon.5;Parent=Bradi1g00213.1.v3.1;pacid=32803704\r\n",
      "Bd1\tphytozomev10\tCDS\t29955\t30309\t.\t-\t0\tID=Bradi1g00213.1.v3.1.CDS.5;Parent=Bradi1g00213.1.v3.1;pacid=32803704\r\n",
      "Bd1\tphytozomev10\texon\t28710\t29935\t.\t-\t.\tID=Bradi1g00213.1.v3.1.exon.6;Parent=Bradi1g00213.1.v3.1;pacid=32803704\r\n",
      "Bd1\tphytozomev10\tCDS\t28710\t29935\t.\t-\t2\tID=Bradi1g00213.1.v3.1.CDS.6;Parent=Bradi1g00213.1.v3.1;pacid=32803704\r\n",
      "Bd1\tphytozomev10\tgene\t47522\t49251\t.\t+\t.\tID=Bradi1g00215.v3.1;Name=Bradi1g00215\r\n",
      "Bd1\tphytozomev10\tmRNA\t47522\t49251\t.\t+\t.\tID=Bradi1g00215.1.v3.1;Name=Bradi1g00215.1;pacid=32805405;longest=1;Parent=Bradi1g00215.v3.1\r\n",
      "Bd1\tphytozomev10\texon\t47522\t47689\t.\t+\t.\tID=Bradi1g00215.1.v3.1.exon.1;Parent=Bradi1g00215.1.v3.1;pacid=32805405\r\n",
      "Bd1\tphytozomev10\tfive_prime_UTR\t47522\t47523\t.\t+\t.\tID=Bradi1g00215.1.v3.1.five_prime_UTR.1;Parent=Bradi1g00215.1.v3.1;pacid=32805405\r\n",
      "Bd1\tphytozomev10\tCDS\t47524\t47689\t.\t+\t0\tID=Bradi1g00215.1.v3.1.CDS.1;Parent=Bradi1g00215.1.v3.1;pacid=32805405\r\n",
      "Bd1\tphytozomev10\texon\t48207\t48268\t.\t+\t.\tID=Bradi1g00215.1.v3.1.exon.2;Parent=Bradi1g00215.1.v3.1;pacid=32805405\r\n",
      "Bd1\tphytozomev10\tCDS\t48207\t48268\t.\t+\t2\tID=Bradi1g00215.1.v3.1.CDS.2;Parent=Bradi1g00215.1.v3.1;pacid=32805405\r\n",
      "Bd1\tphytozomev10\texon\t48362\t48443\t.\t+\t.\tID=Bradi1g00215.1.v3.1.exon.3;Parent=Bradi1g00215.1.v3.1;pacid=32805405\r\n",
      "Bd1\tphytozomev10\tCDS\t48362\t48443\t.\t+\t0\tID=Bradi1g00215.1.v3.1.CDS.3;Parent=Bradi1g00215.1.v3.1;pacid=32805405\r\n",
      "Bd1\tphytozomev10\texon\t48996\t49251\t.\t+\t.\tID=Bradi1g00215.1.v3.1.exon.4;Parent=Bradi1g00215.1.v3.1;pacid=32805405\r\n",
      "Bd1\tphytozomev10\tCDS\t48996\t49003\t.\t+\t2\tID=Bradi1g00215.1.v3.1.CDS.4;Parent=Bradi1g00215.1.v3.1;pacid=32805405\r\n",
      "Bd1\tphytozomev10\tthree_prime_UTR\t49004\t49251\t.\t+\t.\tID=Bradi1g00215.1.v3.1.three_prime_UTR.1;Parent=Bradi1g00215.1.v3.1;pacid=32805405\r\n",
      "Bd1\tphytozomev10\tgene\t52489\t52869\t.\t+\t.\tID=Bradi1g00217.v3.1;Name=Bradi1g00217\r\n",
      "Bd1\tphytozomev10\tmRNA\t52489\t52869\t.\t+\t.\tID=Bradi1g00217.1.v3.1;Name=Bradi1g00217.1;pacid=32794445;longest=1;Parent=Bradi1g00217.v3.1\r\n",
      "Bd1\tphytozomev10\texon\t52489\t52869\t.\t+\t.\tID=Bradi1g00217.1.v3.1.exon.1;Parent=Bradi1g00217.1.v3.1;pacid=32794445\r\n",
      "Bd1\tphytozomev10\tCDS\t52489\t52869\t.\t+\t0\tID=Bradi1g00217.1.v3.1.CDS.1;Parent=Bradi1g00217.1.v3.1;pacid=32794445\r\n",
      "Bd1\tphytozomev10\tgene\t54094\t56379\t.\t+\t.\tID=Bradi1g00219.v3.1;Name=Bradi1g00219;ancestorIdentifier=Bradi1g00219.v2.1\r\n",
      "Bd1\tphytozomev10\tmRNA\t54094\t56379\t.\t+\t.\tID=Bradi1g00219.1.v3.1;Name=Bradi1g00219.1;pacid=32803771;longest=1;ancestorIdentifier=Bradi1g00219.1.v2.1;Parent=Bradi1g00219.v3.1\r\n",
      "Bd1\tphytozomev10\texon\t54094\t54461\t.\t+\t.\tID=Bradi1g00219.1.v3.1.exon.1;Parent=Bradi1g00219.1.v3.1;pacid=32803771\r\n",
      "Bd1\tphytozomev10\tfive_prime_UTR\t54094\t54461\t.\t+\t.\tID=Bradi1g00219.1.v3.1.five_prime_UTR.1;Parent=Bradi1g00219.1.v3.1;pacid=32803771\r\n",
      "Bd1\tphytozomev10\texon\t54558\t54655\t.\t+\t.\tID=Bradi1g00219.1.v3.1.exon.2;Parent=Bradi1g00219.1.v3.1;pacid=32803771\r\n",
      "Bd1\tphytozomev10\tfive_prime_UTR\t54558\t54655\t.\t+\t.\tID=Bradi1g00219.1.v3.1.five_prime_UTR.2;Parent=Bradi1g00219.1.v3.1;pacid=32803771\r\n",
      "Bd1\tphytozomev10\texon\t55006\t56379\t.\t+\t.\tID=Bradi1g00219.1.v3.1.exon.3;Parent=Bradi1g00219.1.v3.1;pacid=32803771\r\n",
      "Bd1\tphytozomev10\tfive_prime_UTR\t55006\t55318\t.\t+\t.\tID=Bradi1g00219.1.v3.1.five_prime_UTR.3;Parent=Bradi1g00219.1.v3.1;pacid=32803771\r\n",
      "Bd1\tphytozomev10\tCDS\t55319\t56140\t.\t+\t0\tID=Bradi1g00219.1.v3.1.CDS.1;Parent=Bradi1g00219.1.v3.1;pacid=32803771\r\n",
      "Bd1\tphytozomev10\tthree_prime_UTR\t56141\t56379\t.\t+\t.\tID=Bradi1g00219.1.v3.1.three_prime_UTR.1;Parent=Bradi1g00219.1.v3.1;pacid=32803771\r\n",
      "Bd1\tphytozomev10\tgene\t56819\t57537\t.\t-\t.\tID=Bradi1g00222.v3.1;Name=Bradi1g00222;ancestorIdentifier=Bradi1g00222.v2.1\r\n",
      "Bd1\tphytozomev10\tmRNA\t56819\t57537\t.\t-\t.\tID=Bradi1g00222.1.v3.1;Name=Bradi1g00222.1;pacid=32802538;longest=1;ancestorIdentifier=Bradi1g00222.1.v2.1;Parent=Bradi1g00222.v3.1\r\n",
      "Bd1\tphytozomev10\texon\t57469\t57537\t.\t-\t.\tID=Bradi1g00222.1.v3.1.exon.1;Parent=Bradi1g00222.1.v3.1;pacid=32802538\r\n",
      "Bd1\tphytozomev10\tCDS\t57469\t57537\t.\t-\t0\tID=Bradi1g00222.1.v3.1.CDS.1;Parent=Bradi1g00222.1.v3.1;pacid=32802538\r\n",
      "Bd1\tphytozomev10\texon\t57254\t57308\t.\t-\t.\tID=Bradi1g00222.1.v3.1.exon.2;Parent=Bradi1g00222.1.v3.1;pacid=32802538\r\n",
      "Bd1\tphytozomev10\tCDS\t57254\t57308\t.\t-\t0\tID=Bradi1g00222.1.v3.1.CDS.2;Parent=Bradi1g00222.1.v3.1;pacid=32802538\r\n",
      "Bd1\tphytozomev10\texon\t56965\t57105\t.\t-\t.\tID=Bradi1g00222.1.v3.1.exon.3;Parent=Bradi1g00222.1.v3.1;pacid=32802538\r\n",
      "Bd1\tphytozomev10\tCDS\t56965\t57105\t.\t-\t2\tID=Bradi1g00222.1.v3.1.CDS.3;Parent=Bradi1g00222.1.v3.1;pacid=32802538\r\n",
      "Bd1\tphytozomev10\texon\t56819\t56877\t.\t-\t.\tID=Bradi1g00222.1.v3.1.exon.4;Parent=Bradi1g00222.1.v3.1;pacid=32802538\r\n",
      "Bd1\tphytozomev10\tCDS\t56819\t56877\t.\t-\t2\tID=Bradi1g00222.1.v3.1.CDS.4;Parent=Bradi1g00222.1.v3.1;pacid=32802538\r\n",
      "Bd1\tphytozomev10\tgene\t59795\t70558\t.\t+\t.\tID=Bradi1g00234.v3.1;Name=Bradi1g00234\r\n",
      "Bd1\tphytozomev10\tmRNA\t59795\t70558\t.\t+\t.\tID=Bradi1g00234.1.v3.1;Name=Bradi1g00234.1;pacid=32794181;longest=1;Parent=Bradi1g00234.v3.1\r\n",
      "Bd1\tphytozomev10\texon\t59795\t61294\t.\t+\t.\tID=Bradi1g00234.1.v3.1.exon.1;Parent=Bradi1g00234.1.v3.1;pacid=32794181\r\n",
      "Bd1\tphytozomev10\tCDS\t59795\t61294\t.\t+\t0\tID=Bradi1g00234.1.v3.1.CDS.1;Parent=Bradi1g00234.1.v3.1;pacid=32794181\r\n",
      "Bd1\tphytozomev10\texon\t63270\t64083\t.\t+\t.\tID=Bradi1g00234.1.v3.1.exon.2;Parent=Bradi1g00234.1.v3.1;pacid=32794181\r\n",
      "Bd1\tphytozomev10\tCDS\t63270\t64083\t.\t+\t0\tID=Bradi1g00234.1.v3.1.CDS.2;Parent=Bradi1g00234.1.v3.1;pacid=32794181\r\n",
      "Bd1\tphytozomev10\texon\t69417\t69898\t.\t+\t.\tID=Bradi1g00234.1.v3.1.exon.3;Parent=Bradi1g00234.1.v3.1;pacid=32794181\r\n",
      "Bd1\tphytozomev10\tCDS\t69417\t69898\t.\t+\t2\tID=Bradi1g00234.1.v3.1.CDS.3;Parent=Bradi1g00234.1.v3.1;pacid=32794181\r\n",
      "Bd1\tphytozomev10\texon\t70007\t70558\t.\t+\t.\tID=Bradi1g00234.1.v3.1.exon.4;Parent=Bradi1g00234.1.v3.1;pacid=32794181\r\n",
      "Bd1\tphytozomev10\tthree_prime_UTR\t70007\t70558\t.\t+\t.\tID=Bradi1g00234.1.v3.1.three_prime_UTR.1;Parent=Bradi1g00234.1.v3.1;pacid=32794181\r\n",
      "Bd1\tphytozomev10\tgene\t104368\t106734\t.\t-\t.\tID=Bradi1g00247.v3.1;Name=Bradi1g00247;ancestorIdentifier=Bradi1g00247.v2.1\r\n",
      "Bd1\tphytozomev10\tmRNA\t104368\t106734\t.\t-\t.\tID=Bradi1g00247.3.v3.1;Name=Bradi1g00247.3;pacid=32796952;longest=1;Parent=Bradi1g00247.v3.1\r\n",
      "Bd1\tphytozomev10\texon\t106306\t106734\t.\t-\t.\tID=Bradi1g00247.3.v3.1.exon.1;Parent=Bradi1g00247.3.v3.1;pacid=32796952\r\n",
      "Bd1\tphytozomev10\tCDS\t106306\t106640\t.\t-\t0\tID=Bradi1g00247.3.v3.1.CDS.1;Parent=Bradi1g00247.3.v3.1;pacid=32796952\r\n",
      "Bd1\tphytozomev10\tfive_prime_UTR\t106641\t106734\t.\t-\t.\tID=Bradi1g00247.3.v3.1.five_prime_UTR.1;Parent=Bradi1g00247.3.v3.1;pacid=32796952\r\n",
      "Bd1\tphytozomev10\texon\t105352\t105448\t.\t-\t.\tID=Bradi1g00247.3.v3.1.exon.2;Parent=Bradi1g00247.3.v3.1;pacid=32796952\r\n"
     ]
    }
   ],
   "source": [
    "!head -n 100 /home/feng/ref/Brachypodium_Bd21_v3-1/annotation/gene_all.gff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##gff-version 3\r\n",
      "##annot-version v3.1\r\n",
      "##species Brachypodium distachyon\r\n",
      "Bd1\tphytozomev10\tgene\t10581\t11638\t.\t+\t.\tID=Bradi1g00200.v3.1;Name=Bradi1g00200;ancestorIdentifier=Bradi1g00200.v2.1\r\n",
      "Bd1\tphytozomev10\tmRNA\t10581\t11638\t.\t+\t.\tID=Bradi1g00200.1.v3.1;Name=Bradi1g00200.1;pacid=32803675;longest=1;ancestorIdentifier=Bradi1g00200.1.v2.1;Parent=Bradi1g00200.v3.1\r\n",
      "Bd1\tphytozomev10\tCDS\t10581\t10850\t.\t+\t0\tID=Bradi1g00200.1.v3.1.CDS.1;Parent=Bradi1g00200.1.v3.1;pacid=32803675\r\n",
      "Bd1\tphytozomev10\tCDS\t11252\t11638\t.\t+\t0\tID=Bradi1g00200.1.v3.1.CDS.2;Parent=Bradi1g00200.1.v3.1;pacid=32803675\r\n",
      "Bd1\tphytozomev10\tgene\t18453\t21742\t.\t+\t.\tID=Bradi1g00210.v3.1;Name=Bradi1g00210;ancestorIdentifier=Bradi1g00210.v2.1\r\n",
      "Bd1\tphytozomev10\tmRNA\t18453\t21742\t.\t+\t.\tID=Bradi1g00210.3.v3.1;Name=Bradi1g00210.3;pacid=32804311;longest=1;Parent=Bradi1g00210.v3.1\r\n",
      "Bd1\tphytozomev10\tfive_prime_UTR\t18453\t18498\t.\t+\t.\tID=Bradi1g00210.3.v3.1.five_prime_UTR.1;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\tCDS\t18499\t18628\t.\t+\t0\tID=Bradi1g00210.3.v3.1.CDS.1;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\tCDS\t18726\t18773\t.\t+\t2\tID=Bradi1g00210.3.v3.1.CDS.2;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\tCDS\t19132\t19176\t.\t+\t2\tID=Bradi1g00210.3.v3.1.CDS.3;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\tCDS\t19327\t19446\t.\t+\t2\tID=Bradi1g00210.3.v3.1.CDS.4;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\tCDS\t19750\t19895\t.\t+\t2\tID=Bradi1g00210.3.v3.1.CDS.5;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\tCDS\t20030\t20177\t.\t+\t0\tID=Bradi1g00210.3.v3.1.CDS.6;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\tCDS\t20599\t20682\t.\t+\t2\tID=Bradi1g00210.3.v3.1.CDS.7;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\tCDS\t20787\t20902\t.\t+\t2\tID=Bradi1g00210.3.v3.1.CDS.8;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\tCDS\t21165\t21223\t.\t+\t0\tID=Bradi1g00210.3.v3.1.CDS.9;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\tCDS\t21335\t21425\t.\t+\t1\tID=Bradi1g00210.3.v3.1.CDS.10;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\tCDS\t21515\t21742\t.\t+\t0\tID=Bradi1g00210.3.v3.1.CDS.11;Parent=Bradi1g00210.3.v3.1;pacid=32804311\r\n",
      "Bd1\tphytozomev10\tgene\t28710\t31462\t.\t-\t.\tID=Bradi1g00213.v3.1;Name=Bradi1g00213;ancestorIdentifier=Bradi1g00213.v2.1\r\n",
      "Bd1\tphytozomev10\tmRNA\t28710\t31462\t.\t-\t.\tID=Bradi1g00213.1.v3.1;Name=Bradi1g00213.1;pacid=32803704;longest=1;ancestorIdentifier=Bradi1g00213.1.v2.1;Parent=Bradi1g00213.v3.1\r\n",
      "Bd1\tphytozomev10\tCDS\t31315\t31462\t.\t-\t0\tID=Bradi1g00213.1.v3.1.CDS.1;Parent=Bradi1g00213.1.v3.1;pacid=32803704\r\n",
      "Bd1\tphytozomev10\tCDS\t31064\t31132\t.\t-\t2\tID=Bradi1g00213.1.v3.1.CDS.2;Parent=Bradi1g00213.1.v3.1;pacid=32803704\r\n",
      "Bd1\tphytozomev10\tCDS\t30640\t31013\t.\t-\t2\tID=Bradi1g00213.1.v3.1.CDS.3;Parent=Bradi1g00213.1.v3.1;pacid=32803704\r\n",
      "Bd1\tphytozomev10\tCDS\t30412\t30570\t.\t-\t0\tID=Bradi1g00213.1.v3.1.CDS.4;Parent=Bradi1g00213.1.v3.1;pacid=32803704\r\n",
      "Bd1\tphytozomev10\tCDS\t29955\t30309\t.\t-\t0\tID=Bradi1g00213.1.v3.1.CDS.5;Parent=Bradi1g00213.1.v3.1;pacid=32803704\r\n",
      "Bd1\tphytozomev10\tCDS\t28710\t29935\t.\t-\t2\tID=Bradi1g00213.1.v3.1.CDS.6;Parent=Bradi1g00213.1.v3.1;pacid=32803704\r\n",
      "Bd1\tphytozomev10\tgene\t47522\t49251\t.\t+\t.\tID=Bradi1g00215.v3.1;Name=Bradi1g00215\r\n",
      "Bd1\tphytozomev10\tmRNA\t47522\t49251\t.\t+\t.\tID=Bradi1g00215.1.v3.1;Name=Bradi1g00215.1;pacid=32805405;longest=1;Parent=Bradi1g00215.v3.1\r\n",
      "Bd1\tphytozomev10\tfive_prime_UTR\t47522\t47523\t.\t+\t.\tID=Bradi1g00215.1.v3.1.five_prime_UTR.1;Parent=Bradi1g00215.1.v3.1;pacid=32805405\r\n",
      "Bd1\tphytozomev10\tCDS\t47524\t47689\t.\t+\t0\tID=Bradi1g00215.1.v3.1.CDS.1;Parent=Bradi1g00215.1.v3.1;pacid=32805405\r\n",
      "Bd1\tphytozomev10\tCDS\t48207\t48268\t.\t+\t2\tID=Bradi1g00215.1.v3.1.CDS.2;Parent=Bradi1g00215.1.v3.1;pacid=32805405\r\n",
      "Bd1\tphytozomev10\tCDS\t48362\t48443\t.\t+\t0\tID=Bradi1g00215.1.v3.1.CDS.3;Parent=Bradi1g00215.1.v3.1;pacid=32805405\r\n",
      "Bd1\tphytozomev10\tCDS\t48996\t49003\t.\t+\t2\tID=Bradi1g00215.1.v3.1.CDS.4;Parent=Bradi1g00215.1.v3.1;pacid=32805405\r\n",
      "Bd1\tphytozomev10\tthree_prime_UTR\t49004\t49251\t.\t+\t.\tID=Bradi1g00215.1.v3.1.three_prime_UTR.1;Parent=Bradi1g00215.1.v3.1;pacid=32805405\r\n",
      "Bd1\tphytozomev10\tgene\t52489\t52869\t.\t+\t.\tID=Bradi1g00217.v3.1;Name=Bradi1g00217\r\n",
      "Bd1\tphytozomev10\tmRNA\t52489\t52869\t.\t+\t.\tID=Bradi1g00217.1.v3.1;Name=Bradi1g00217.1;pacid=32794445;longest=1;Parent=Bradi1g00217.v3.1\r\n",
      "Bd1\tphytozomev10\tCDS\t52489\t52869\t.\t+\t0\tID=Bradi1g00217.1.v3.1.CDS.1;Parent=Bradi1g00217.1.v3.1;pacid=32794445\r\n",
      "Bd1\tphytozomev10\tgene\t54094\t56379\t.\t+\t.\tID=Bradi1g00219.v3.1;Name=Bradi1g00219;ancestorIdentifier=Bradi1g00219.v2.1\r\n",
      "Bd1\tphytozomev10\tmRNA\t54094\t56379\t.\t+\t.\tID=Bradi1g00219.1.v3.1;Name=Bradi1g00219.1;pacid=32803771;longest=1;ancestorIdentifier=Bradi1g00219.1.v2.1;Parent=Bradi1g00219.v3.1\r\n",
      "Bd1\tphytozomev10\tfive_prime_UTR\t54094\t54461\t.\t+\t.\tID=Bradi1g00219.1.v3.1.five_prime_UTR.1;Parent=Bradi1g00219.1.v3.1;pacid=32803771\r\n",
      "Bd1\tphytozomev10\tfive_prime_UTR\t54558\t54655\t.\t+\t.\tID=Bradi1g00219.1.v3.1.five_prime_UTR.2;Parent=Bradi1g00219.1.v3.1;pacid=32803771\r\n",
      "Bd1\tphytozomev10\tfive_prime_UTR\t55006\t55318\t.\t+\t.\tID=Bradi1g00219.1.v3.1.five_prime_UTR.3;Parent=Bradi1g00219.1.v3.1;pacid=32803771\r\n",
      "Bd1\tphytozomev10\tCDS\t55319\t56140\t.\t+\t0\tID=Bradi1g00219.1.v3.1.CDS.1;Parent=Bradi1g00219.1.v3.1;pacid=32803771\r\n",
      "Bd1\tphytozomev10\tthree_prime_UTR\t56141\t56379\t.\t+\t.\tID=Bradi1g00219.1.v3.1.three_prime_UTR.1;Parent=Bradi1g00219.1.v3.1;pacid=32803771\r\n",
      "Bd1\tphytozomev10\tgene\t56819\t57537\t.\t-\t.\tID=Bradi1g00222.v3.1;Name=Bradi1g00222;ancestorIdentifier=Bradi1g00222.v2.1\r\n",
      "Bd1\tphytozomev10\tmRNA\t56819\t57537\t.\t-\t.\tID=Bradi1g00222.1.v3.1;Name=Bradi1g00222.1;pacid=32802538;longest=1;ancestorIdentifier=Bradi1g00222.1.v2.1;Parent=Bradi1g00222.v3.1\r\n",
      "Bd1\tphytozomev10\tCDS\t57469\t57537\t.\t-\t0\tID=Bradi1g00222.1.v3.1.CDS.1;Parent=Bradi1g00222.1.v3.1;pacid=32802538\r\n",
      "Bd1\tphytozomev10\tCDS\t57254\t57308\t.\t-\t0\tID=Bradi1g00222.1.v3.1.CDS.2;Parent=Bradi1g00222.1.v3.1;pacid=32802538\r\n",
      "Bd1\tphytozomev10\tCDS\t56965\t57105\t.\t-\t2\tID=Bradi1g00222.1.v3.1.CDS.3;Parent=Bradi1g00222.1.v3.1;pacid=32802538\r\n",
      "Bd1\tphytozomev10\tCDS\t56819\t56877\t.\t-\t2\tID=Bradi1g00222.1.v3.1.CDS.4;Parent=Bradi1g00222.1.v3.1;pacid=32802538\r\n",
      "Bd1\tphytozomev10\tgene\t59795\t70558\t.\t+\t.\tID=Bradi1g00234.v3.1;Name=Bradi1g00234\r\n",
      "Bd1\tphytozomev10\tmRNA\t59795\t70558\t.\t+\t.\tID=Bradi1g00234.1.v3.1;Name=Bradi1g00234.1;pacid=32794181;longest=1;Parent=Bradi1g00234.v3.1\r\n",
      "Bd1\tphytozomev10\tCDS\t59795\t61294\t.\t+\t0\tID=Bradi1g00234.1.v3.1.CDS.1;Parent=Bradi1g00234.1.v3.1;pacid=32794181\r\n",
      "Bd1\tphytozomev10\tCDS\t63270\t64083\t.\t+\t0\tID=Bradi1g00234.1.v3.1.CDS.2;Parent=Bradi1g00234.1.v3.1;pacid=32794181\r\n",
      "Bd1\tphytozomev10\tCDS\t69417\t69898\t.\t+\t2\tID=Bradi1g00234.1.v3.1.CDS.3;Parent=Bradi1g00234.1.v3.1;pacid=32794181\r\n",
      "Bd1\tphytozomev10\tthree_prime_UTR\t70007\t70558\t.\t+\t.\tID=Bradi1g00234.1.v3.1.three_prime_UTR.1;Parent=Bradi1g00234.1.v3.1;pacid=32794181\r\n",
      "Bd1\tphytozomev10\tgene\t104368\t106734\t.\t-\t.\tID=Bradi1g00247.v3.1;Name=Bradi1g00247;ancestorIdentifier=Bradi1g00247.v2.1\r\n",
      "Bd1\tphytozomev10\tmRNA\t104368\t106734\t.\t-\t.\tID=Bradi1g00247.3.v3.1;Name=Bradi1g00247.3;pacid=32796952;longest=1;Parent=Bradi1g00247.v3.1\r\n",
      "Bd1\tphytozomev10\tCDS\t106306\t106640\t.\t-\t0\tID=Bradi1g00247.3.v3.1.CDS.1;Parent=Bradi1g00247.3.v3.1;pacid=32796952\r\n",
      "Bd1\tphytozomev10\tfive_prime_UTR\t106641\t106734\t.\t-\t.\tID=Bradi1g00247.3.v3.1.five_prime_UTR.1;Parent=Bradi1g00247.3.v3.1;pacid=32796952\r\n",
      "Bd1\tphytozomev10\tCDS\t105352\t105448\t.\t-\t1\tID=Bradi1g00247.3.v3.1.CDS.2;Parent=Bradi1g00247.3.v3.1;pacid=32796952\r\n",
      "Bd1\tphytozomev10\tCDS\t105171\t105230\t.\t-\t0\tID=Bradi1g00247.3.v3.1.CDS.3;Parent=Bradi1g00247.3.v3.1;pacid=32796952\r\n",
      "Bd1\tphytozomev10\tthree_prime_UTR\t104368\t104771\t.\t-\t.\tID=Bradi1g00247.3.v3.1.three_prime_UTR.1;Parent=Bradi1g00247.3.v3.1;pacid=32796952\r\n",
      "Bd1\tphytozomev10\tCDS\t104772\t104852\t.\t-\t0\tID=Bradi1g00247.3.v3.1.CDS.4;Parent=Bradi1g00247.3.v3.1;pacid=32796952\r\n",
      "Bd1\tphytozomev10\tgene\t108527\t109357\t.\t+\t.\tID=Bradi1g00253.v3.1;Name=Bradi1g00253;ancestorIdentifier=Bradi1g00253.v2.1\r\n",
      "Bd1\tphytozomev10\tmRNA\t108527\t109357\t.\t+\t.\tID=Bradi1g00253.1.v3.1;Name=Bradi1g00253.1;pacid=32797237;longest=1;ancestorIdentifier=Bradi1g00253.1.v2.1;Parent=Bradi1g00253.v3.1\r\n",
      "Bd1\tphytozomev10\tfive_prime_UTR\t108527\t108781\t.\t+\t.\tID=Bradi1g00253.1.v3.1.five_prime_UTR.1;Parent=Bradi1g00253.1.v3.1;pacid=32797237\r\n",
      "Bd1\tphytozomev10\tCDS\t108782\t109357\t.\t+\t0\tID=Bradi1g00253.1.v3.1.CDS.1;Parent=Bradi1g00253.1.v3.1;pacid=32797237\r\n",
      "Bd1\tphytozomev10\tgene\t112000\t121797\t.\t+\t.\tID=Bradi1g00260.v3.1;Name=Bradi1g00260;ancestorIdentifier=Bradi1g00260.v2.1\r\n",
      "Bd1\tphytozomev10\tmRNA\t112000\t121797\t.\t+\t.\tID=Bradi1g00260.2.v3.1;Name=Bradi1g00260.2;pacid=32797833;longest=1;ancestorIdentifier=Bradi1g00260.2.v2.1;Parent=Bradi1g00260.v3.1\r\n",
      "Bd1\tphytozomev10\tfive_prime_UTR\t112000\t112523\t.\t+\t.\tID=Bradi1g00260.2.v3.1.five_prime_UTR.1;Parent=Bradi1g00260.2.v3.1;pacid=32797833\r\n",
      "Bd1\tphytozomev10\tfive_prime_UTR\t113654\t113998\t.\t+\t.\tID=Bradi1g00260.2.v3.1.five_prime_UTR.2;Parent=Bradi1g00260.2.v3.1;pacid=32797833\r\n",
      "Bd1\tphytozomev10\tfive_prime_UTR\t114167\t114703\t.\t+\t.\tID=Bradi1g00260.2.v3.1.five_prime_UTR.3;Parent=Bradi1g00260.2.v3.1;pacid=32797833\r\n",
      "Bd1\tphytozomev10\tCDS\t114704\t115022\t.\t+\t0\tID=Bradi1g00260.2.v3.1.CDS.1;Parent=Bradi1g00260.2.v3.1;pacid=32797833\r\n",
      "Bd1\tphytozomev10\tCDS\t116199\t116778\t.\t+\t2\tID=Bradi1g00260.2.v3.1.CDS.2;Parent=Bradi1g00260.2.v3.1;pacid=32797833\r\n",
      "Bd1\tphytozomev10\tCDS\t116874\t117364\t.\t+\t1\tID=Bradi1g00260.2.v3.1.CDS.3;Parent=Bradi1g00260.2.v3.1;pacid=32797833\r\n",
      "Bd1\tphytozomev10\tCDS\t117868\t118006\t.\t+\t2\tID=Bradi1g00260.2.v3.1.CDS.4;Parent=Bradi1g00260.2.v3.1;pacid=32797833\r\n",
      "Bd1\tphytozomev10\tCDS\t118141\t118212\t.\t+\t1\tID=Bradi1g00260.2.v3.1.CDS.5;Parent=Bradi1g00260.2.v3.1;pacid=32797833\r\n",
      "Bd1\tphytozomev10\tCDS\t118835\t118906\t.\t+\t1\tID=Bradi1g00260.2.v3.1.CDS.6;Parent=Bradi1g00260.2.v3.1;pacid=32797833\r\n",
      "Bd1\tphytozomev10\tCDS\t118993\t119073\t.\t+\t1\tID=Bradi1g00260.2.v3.1.CDS.7;Parent=Bradi1g00260.2.v3.1;pacid=32797833\r\n",
      "Bd1\tphytozomev10\tCDS\t119345\t119511\t.\t+\t1\tID=Bradi1g00260.2.v3.1.CDS.8;Parent=Bradi1g00260.2.v3.1;pacid=32797833\r\n",
      "Bd1\tphytozomev10\tCDS\t119631\t119833\t.\t+\t2\tID=Bradi1g00260.2.v3.1.CDS.9;Parent=Bradi1g00260.2.v3.1;pacid=32797833\r\n",
      "Bd1\tphytozomev10\tCDS\t119936\t120062\t.\t+\t0\tID=Bradi1g00260.2.v3.1.CDS.10;Parent=Bradi1g00260.2.v3.1;pacid=32797833\r\n",
      "Bd1\tphytozomev10\tCDS\t120453\t120509\t.\t+\t2\tID=Bradi1g00260.2.v3.1.CDS.11;Parent=Bradi1g00260.2.v3.1;pacid=32797833\r\n",
      "Bd1\tphytozomev10\tCDS\t120638\t120827\t.\t+\t2\tID=Bradi1g00260.2.v3.1.CDS.12;Parent=Bradi1g00260.2.v3.1;pacid=32797833\r\n",
      "Bd1\tphytozomev10\tCDS\t120915\t121350\t.\t+\t1\tID=Bradi1g00260.2.v3.1.CDS.13;Parent=Bradi1g00260.2.v3.1;pacid=32797833\r\n",
      "Bd1\tphytozomev10\tthree_prime_UTR\t121351\t121797\t.\t+\t.\tID=Bradi1g00260.2.v3.1.three_prime_UTR.1;Parent=Bradi1g00260.2.v3.1;pacid=32797833\r\n",
      "Bd1\tphytozomev10\tmRNA\t112000\t121797\t.\t+\t.\tID=Bradi1g00260.6.v3.1;Name=Bradi1g00260.6;pacid=32797834;longest=0;Parent=Bradi1g00260.v3.1\r\n",
      "Bd1\tphytozomev10\tfive_prime_UTR\t112000\t114703\t.\t+\t.\tID=Bradi1g00260.6.v3.1.five_prime_UTR.1;Parent=Bradi1g00260.6.v3.1;pacid=32797834\r\n",
      "Bd1\tphytozomev10\tCDS\t114704\t114869\t.\t+\t0\tID=Bradi1g00260.6.v3.1.CDS.1;Parent=Bradi1g00260.6.v3.1;pacid=32797834\r\n",
      "Bd1\tphytozomev10\tCDS\t114942\t115022\t.\t+\t2\tID=Bradi1g00260.6.v3.1.CDS.2;Parent=Bradi1g00260.6.v3.1;pacid=32797834\r\n",
      "Bd1\tphytozomev10\tCDS\t116199\t116778\t.\t+\t2\tID=Bradi1g00260.6.v3.1.CDS.3;Parent=Bradi1g00260.6.v3.1;pacid=32797834\r\n",
      "Bd1\tphytozomev10\tCDS\t116874\t117364\t.\t+\t1\tID=Bradi1g00260.6.v3.1.CDS.4;Parent=Bradi1g00260.6.v3.1;pacid=32797834\r\n",
      "Bd1\tphytozomev10\tCDS\t117868\t118006\t.\t+\t2\tID=Bradi1g00260.6.v3.1.CDS.5;Parent=Bradi1g00260.6.v3.1;pacid=32797834\r\n",
      "Bd1\tphytozomev10\tCDS\t118141\t118212\t.\t+\t1\tID=Bradi1g00260.6.v3.1.CDS.6;Parent=Bradi1g00260.6.v3.1;pacid=32797834\r\n",
      "Bd1\tphytozomev10\tCDS\t118835\t118906\t.\t+\t1\tID=Bradi1g00260.6.v3.1.CDS.7;Parent=Bradi1g00260.6.v3.1;pacid=32797834\r\n",
      "Bd1\tphytozomev10\tCDS\t118993\t119073\t.\t+\t1\tID=Bradi1g00260.6.v3.1.CDS.8;Parent=Bradi1g00260.6.v3.1;pacid=32797834\r\n"
     ]
    }
   ],
   "source": [
    "! head -n 100 /home/feng/ref/Brachypodium_Bd21_v3-1/annotation/Bdistachyon_314_v3.1.gene.gff3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash \n",
    "ln -s /home/feng/ref/Brachypodium_Bd21_v3-1/transcriptome/gene_index.gff /home/feng/ref/Brachypodium_Bd21_v3-1/annotation/gene_all.gff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat /home/feng/ref/Brachypodium_Bd21_v3-1/transcriptome/gene_index.gff | grep gene >/home/feng/ref/Brachypodium_Bd21_v3-1/annotation/gene_only.gff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0       468\n",
       "1       311\n",
       "2       831\n",
       "3       353\n",
       "4       508\n",
       "5       424\n",
       "6       304\n",
       "7       553\n",
       "8       531\n",
       "9       711\n",
       "10      644\n",
       "11      592\n",
       "12      965\n",
       "13      914\n",
       "14      605\n",
       "15      750\n",
       "16      339\n",
       "17      299\n",
       "18      487\n",
       "19      630\n",
       "20      577\n",
       "21      545\n",
       "22      389\n",
       "23      538\n",
       "24      299\n",
       "25      884\n",
       "26      441\n",
       "27      483\n",
       "28      299\n",
       "29      323\n",
       "       ... \n",
       "479     872\n",
       "480     592\n",
       "481     495\n",
       "482     315\n",
       "483     537\n",
       "484     382\n",
       "485     299\n",
       "486     468\n",
       "487     518\n",
       "488    1037\n",
       "489     323\n",
       "490     304\n",
       "491    1276\n",
       "492    1630\n",
       "493     663\n",
       "494     458\n",
       "495    1853\n",
       "496     299\n",
       "497     741\n",
       "498     318\n",
       "499     351\n",
       "500     678\n",
       "501     299\n",
       "502     520\n",
       "503     668\n",
       "504     512\n",
       "505     518\n",
       "506     299\n",
       "507     596\n",
       "508     572\n",
       "dtype: int64"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[2]-df[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "https://github.com/taoliu/MACS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: macs2 callpeak [-h] -t TFILE [TFILE ...] [-c [CFILE [CFILE ...]]]\r\n",
      "                      [-f {AUTO,BAM,SAM,BED,ELAND,ELANDMULTI,ELANDEXPORT,BOWTIE,BAMPE}]\r\n",
      "                      [-g GSIZE] [--keep-dup KEEPDUPLICATES]\r\n",
      "                      [--buffer-size BUFFER_SIZE] [--outdir OUTDIR] [-n NAME]\r\n",
      "                      [-B] [--verbose VERBOSE] [--trackline] [--SPMR]\r\n",
      "                      [-s TSIZE] [--bw BW] [-m MFOLD MFOLD] [--fix-bimodal]\r\n",
      "                      [--nomodel] [--shift SHIFT] [--extsize EXTSIZE]\r\n",
      "                      [-q QVALUE | -p PVALUE] [--to-large] [--ratio RATIO]\r\n",
      "                      [--down-sample] [--seed SEED] [--nolambda]\r\n",
      "                      [--slocal SMALLLOCAL] [--llocal LARGELOCAL] [--broad]\r\n",
      "                      [--broad-cutoff BROADCUTOFF] [--cutoff-analysis]\r\n",
      "                      [--call-summits] [--fe-cutoff FECUTOFF]\r\n",
      "\r\n",
      "optional arguments:\r\n",
      "  -h, --help            show this help message and exit\r\n",
      "\r\n",
      "Input files arguments:\r\n",
      "  -t TFILE [TFILE ...], --treatment TFILE [TFILE ...]\r\n",
      "                        ChIP-seq treatment file. If multiple files are given\r\n",
      "                        as '-t A B C', then they will all be read and pooled\r\n",
      "                        together. REQUIRED.\r\n",
      "  -c [CFILE [CFILE ...]], --control [CFILE [CFILE ...]]\r\n",
      "                        Control file. If multiple files are given as '-c A B\r\n",
      "                        C', they will be pooled to estimate ChIP-seq\r\n",
      "                        background noise.\r\n",
      "  -f {AUTO,BAM,SAM,BED,ELAND,ELANDMULTI,ELANDEXPORT,BOWTIE,BAMPE}, --format {AUTO,BAM,SAM,BED,ELAND,ELANDMULTI,ELANDEXPORT,BOWTIE,BAMPE}\r\n",
      "                        Format of tag file, \"AUTO\", \"BED\" or \"ELAND\" or\r\n",
      "                        \"ELANDMULTI\" or \"ELANDEXPORT\" or \"SAM\" or \"BAM\" or\r\n",
      "                        \"BOWTIE\" or \"BAMPE\". The default AUTO option will let\r\n",
      "                        MACS decide which format the file is. Please check the\r\n",
      "                        definition in README file if you choose\r\n",
      "                        ELAND/ELANDMULTI/ELANDEXPORT/SAM/BAM/BOWTIE. DEFAULT:\r\n",
      "                        \"AUTO\"\r\n",
      "  -g GSIZE, --gsize GSIZE\r\n",
      "                        Effective genome size. It can be 1.0e+9 or 1000000000,\r\n",
      "                        or shortcuts:'hs' for human (2.7e9), 'mm' for mouse\r\n",
      "                        (1.87e9), 'ce' for C. elegans (9e7) and 'dm' for\r\n",
      "                        fruitfly (1.2e8), Default:hs\r\n",
      "  --keep-dup KEEPDUPLICATES\r\n",
      "                        It controls the MACS behavior towards duplicate tags\r\n",
      "                        at the exact same location -- the same coordination\r\n",
      "                        and the same strand. The 'auto' option makes MACS\r\n",
      "                        calculate the maximum tags at the exact same location\r\n",
      "                        based on binomal distribution using 1e-5 as pvalue\r\n",
      "                        cutoff; and the 'all' option keeps every tags. If an\r\n",
      "                        integer is given, at most this number of tags will be\r\n",
      "                        kept at the same location. The default is to keep one\r\n",
      "                        tag at the same location. Default: 1\r\n",
      "  --buffer-size BUFFER_SIZE\r\n",
      "                        Buffer size for incrementally increasing internal\r\n",
      "                        array size to store reads alignment information. In\r\n",
      "                        most cases, you don't have to change this parameter.\r\n",
      "                        However, if there are large number of\r\n",
      "                        chromosomes/contigs/scaffolds in your alignment, it's\r\n",
      "                        recommended to specify a smaller buffer size in order\r\n",
      "                        to decrease memory usage (but it will take longer time\r\n",
      "                        to read alignment files). Minimum memory requested for\r\n",
      "                        reading an alignment file is about # of CHROMOSOME *\r\n",
      "                        BUFFER_SIZE * 2 Bytes. DEFAULT: 100000\r\n",
      "\r\n",
      "Output arguments:\r\n",
      "  --outdir OUTDIR       If specified all output files will be written to that\r\n",
      "                        directory. Default: the current working directory\r\n",
      "  -n NAME, --name NAME  Experiment name, which will be used to generate output\r\n",
      "                        file names. DEFAULT: \"NA\"\r\n",
      "  -B, --bdg             Whether or not to save extended fragment pileup, and\r\n",
      "                        local lambda tracks (two files) at every bp into a\r\n",
      "                        bedGraph file. DEFAULT: False\r\n",
      "  --verbose VERBOSE     Set verbose level of runtime message. 0: only show\r\n",
      "                        critical message, 1: show additional warning message,\r\n",
      "                        2: show process information, 3: show debug messages.\r\n",
      "                        DEFAULT:2\r\n",
      "  --trackline           Tells MACS to include trackline with bedGraph files.\r\n",
      "                        To include this trackline while displaying bedGraph at\r\n",
      "                        UCSC genome browser, can show name and description of\r\n",
      "                        the file as well. However my suggestion is to convert\r\n",
      "                        bedGraph to bigWig, then show the smaller and faster\r\n",
      "                        binary bigWig file at UCSC genome browser, as well as\r\n",
      "                        downstream analysis. Require -B to be set. Default:\r\n",
      "                        Not include trackline.\r\n",
      "  --SPMR                If True, MACS will save signal per million reads for\r\n",
      "                        fragment pileup profiles. Require -B to be set.\r\n",
      "                        Default: False\r\n",
      "\r\n",
      "Shifting model arguments:\r\n",
      "  -s TSIZE, --tsize TSIZE\r\n",
      "                        Tag size. This will overide the auto detected tag\r\n",
      "                        size. DEFAULT: Not set\r\n",
      "  --bw BW               Band width for picking regions to compute fragment\r\n",
      "                        size. This value is only used while building the\r\n",
      "                        shifting model. DEFAULT: 300\r\n",
      "  -m MFOLD MFOLD, --mfold MFOLD MFOLD\r\n",
      "                        Select the regions within MFOLD range of high-\r\n",
      "                        confidence enrichment ratio against background to\r\n",
      "                        build model. Fold-enrichment in regions must be lower\r\n",
      "                        than upper limit, and higher than the lower limit. Use\r\n",
      "                        as \"-m 10 30\". DEFAULT:5 50\r\n",
      "  --fix-bimodal         Whether turn on the auto pair model process. If set,\r\n",
      "                        when MACS failed to build paired model, it will use\r\n",
      "                        the nomodel settings, the --exsize parameter to extend\r\n",
      "                        each tags towards 3' direction. Not to use this\r\n",
      "                        automate fixation is a default behavior now. DEFAULT:\r\n",
      "                        False\r\n",
      "  --nomodel             Whether or not to build the shifting model. If True,\r\n",
      "                        MACS will not build model. by default it means\r\n",
      "                        shifting size = 100, try to set extsize to change it.\r\n",
      "                        DEFAULT: False\r\n",
      "  --shift SHIFT         (NOT the legacy --shiftsize option!) The arbitrary\r\n",
      "                        shift in bp. Use discretion while setting it other\r\n",
      "                        than default value. When NOMODEL is set, MACS will use\r\n",
      "                        this value to move cutting ends (5') towards 5'->3'\r\n",
      "                        direction then apply EXTSIZE to extend them to\r\n",
      "                        fragments. When this value is negative, ends will be\r\n",
      "                        moved toward 3'->5' direction. Recommended to keep it\r\n",
      "                        as default 0 for ChIP-Seq datasets, or -1 * half of\r\n",
      "                        EXTSIZE together with EXTSIZE option for detecting\r\n",
      "                        enriched cutting loci such as certain DNAseI-Seq\r\n",
      "                        datasets. Note, you can't set values other than 0 if\r\n",
      "                        format is BAMPE for paired-end data. DEFAULT: 0.\r\n",
      "  --extsize EXTSIZE     The arbitrary extension size in bp. When nomodel is\r\n",
      "                        true, MACS will use this value as fragment size to\r\n",
      "                        extend each read towards 3' end, then pile them up.\r\n",
      "                        It's exactly twice the number of obsolete SHIFTSIZE.\r\n",
      "                        In previous language, each read is moved 5'->3'\r\n",
      "                        direction to middle of fragment by 1/2 d, then\r\n",
      "                        extended to both direction with 1/2 d. This is\r\n",
      "                        equivalent to say each read is extended towards 5'->3'\r\n",
      "                        into a d size fragment. DEFAULT: 200. EXTSIZE and\r\n",
      "                        SHIFT can be combined when necessary. Check SHIFT\r\n",
      "                        option.\r\n",
      "\r\n",
      "Peak calling arguments:\r\n",
      "  -q QVALUE, --qvalue QVALUE\r\n",
      "                        Minimum FDR (q-value) cutoff for peak detection.\r\n",
      "                        DEFAULT: 0.05. -q, and -p are mutually exclusive.\r\n",
      "  -p PVALUE, --pvalue PVALUE\r\n",
      "                        Pvalue cutoff for peak detection. DEFAULT: not set.\r\n",
      "                        -q, and -p are mutually exclusive. If pvalue cutoff is\r\n",
      "                        set, qvalue will not be calculated and reported as -1\r\n",
      "                        in the final .xls file.\r\n",
      "  --to-large            When set, scale the small sample up to the bigger\r\n",
      "                        sample. By default, the bigger dataset will be scaled\r\n",
      "                        down towards the smaller dataset, which will lead to\r\n",
      "                        smaller p/qvalues and more specific results. Keep in\r\n",
      "                        mind that scaling down will bring down background\r\n",
      "                        noise more. DEFAULT: False\r\n",
      "  --ratio RATIO         When set, use a custom scaling ratio of ChIP/control\r\n",
      "                        (e.g. calculated using NCIS) for linear scaling.\r\n",
      "                        DEFAULT: ingore\r\n",
      "  --down-sample         When set, random sampling method will scale down the\r\n",
      "                        bigger sample. By default, MACS uses linear scaling.\r\n",
      "                        Warning: This option will make your result unstable\r\n",
      "                        and irreproducible since each time, random reads would\r\n",
      "                        be selected. Consider to use 'randsample' script\r\n",
      "                        instead. <not implmented>If used together with --SPMR,\r\n",
      "                        1 million unique reads will be randomly picked.</not\r\n",
      "                        implemented> Caution: due to the implementation, the\r\n",
      "                        final number of selected reads may not be as you\r\n",
      "                        expected! DEFAULT: False\r\n",
      "  --seed SEED           Set the random seed while down sampling data. Must be\r\n",
      "                        a non-negative integer in order to be effective.\r\n",
      "                        DEFAULT: not set\r\n",
      "  --nolambda            If True, MACS will use fixed background lambda as\r\n",
      "                        local lambda for every peak region. Normally, MACS\r\n",
      "                        calculates a dynamic local lambda to reflect the local\r\n",
      "                        bias due to potential chromatin structure.\r\n",
      "  --slocal SMALLLOCAL   The small nearby region in basepairs to calculate\r\n",
      "                        dynamic lambda. This is used to capture the bias near\r\n",
      "                        the peak summit region. Invalid if there is no control\r\n",
      "                        data. If you set this to 0, MACS will skip slocal\r\n",
      "                        lambda calculation. *Note* that MACS will always\r\n",
      "                        perform a d-size local lambda calculation. The final\r\n",
      "                        local bias should be the maximum of the lambda value\r\n",
      "                        from d, slocal, and llocal size windows. DEFAULT: 1000\r\n",
      "  --llocal LARGELOCAL   The large nearby region in basepairs to calculate\r\n",
      "                        dynamic lambda. This is used to capture the surround\r\n",
      "                        bias. If you set this to 0, MACS will skip llocal\r\n",
      "                        lambda calculation. *Note* that MACS will always\r\n",
      "                        perform a d-size local lambda calculation. The final\r\n",
      "                        local bias should be the maximum of the lambda value\r\n",
      "                        from d, slocal, and llocal size windows. DEFAULT:\r\n",
      "                        10000.\r\n",
      "  --broad               If set, MACS will try to call broad peaks by linking\r\n",
      "                        nearby highly enriched regions. The linking region is\r\n",
      "                        controlled by another cutoff through --linking-cutoff.\r\n",
      "                        The maximum linking region length is 4 times of d from\r\n",
      "                        MACS. DEFAULT: False\r\n",
      "  --broad-cutoff BROADCUTOFF\r\n",
      "                        Cutoff for broad region. This option is not available\r\n",
      "                        unless --broad is set. If -p is set, this is a pvalue\r\n",
      "                        cutoff, otherwise, it's a qvalue cutoff. DEFAULT: 0.1\r\n",
      "  --cutoff-analysis     While set, MACS2 will analyze number or total length\r\n",
      "                        of peaks that can be called by different p-value\r\n",
      "                        cutoff then output a summary table to help user decide\r\n",
      "                        a better cutoff. The table will be saved in\r\n",
      "                        NAME_cutoff_analysis.txt file. Note, minlen and maxgap\r\n",
      "                        may affect the results. WARNING: May take ~30 folds\r\n",
      "                        longer time to finish. DEFAULT: False\r\n",
      "\r\n",
      "Post-processing options:\r\n",
      "  --call-summits        If set, MACS will use a more sophisticated signal\r\n",
      "                        processing approach to find subpeak summits in each\r\n",
      "                        enriched peak region. DEFAULT: False\r\n",
      "  --fe-cutoff FECUTOFF  When set, the value will be used to filter out peaks\r\n",
      "                        with low fold-enrichment. Note, MACS2 use 1.0 as\r\n",
      "                        pseudocount while calculating fold-enrichment.\r\n",
      "                        DEFAULT: 1.0\r\n"
     ]
    }
   ],
   "source": [
    "! cat HELP"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
