{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Central question:\n",
    "\n",
    "How is flowering in Brachypodium controlled?\n",
    "\n",
    "## Data: \n",
    "\n",
    " 1. RNASeq with a long-day/short-day/mutant contrast, with a circadian coordinate\n",
    " 1. ChIPSeq of ELF3, the transcription factor relaying photoperiod signal from cytochrome c to downstream.\n",
    " \n",
    "## References: \n",
    "\n",
    "1. [Phytochrome c mutant has delayed flowering ( brachypodia )](http://www.genetics.org/content/198/1/397.long)\n",
    "\n",
    "1. [Evening complex in A.thaliana on senseing thermo-signal and photoperiodic signal](https://www.ncbi.nlm.nih.gov/pubmed/27594171)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "FlowChart (Based on FuXiang's slides):\n",
    "\n",
    "1. Mapping reads and QC\n",
    "\n",
    "  * HISAT + StringTie  for RNA-Seq\n",
    "  * bowtie2 for ChIP-Seq\n",
    "  * Metrics: (Coverage and mapped percentage, Assembly version)\n",
    "  * QC: correlation plots (deeptools), PCA (R?)\n",
    "\n",
    "1. Differential expression \n",
    "  * Contrast level: LD/SD/mutant\n",
    "  \n",
    "1. Epigenetics\n",
    "  * MACS2 for peak calling\n",
    "  * look for motif enrichment HOMER, MEME?\n",
    "\n",
    "1. Constrast RNASeq and ChIPSeq result\n",
    "  * Gene ontology and pathway enrichment of the derived gene set.\n",
    "  \n",
    "1. Clustering to obtain coexpression clades.\n",
    "  * Exploit the ZT coordinate\n",
    "  * CLUST/HC\n",
    "  \n",
    "1. Further analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### May.14th\n",
    "\n",
    "Aims:\n",
    "\n",
    "1. [ ]  Setup production environment\n",
    "\n",
    "1. [ ] Gather and label raw data (.fq)\n",
    "\n",
    "1. [ ] Build up and test the pipeline for HISAT+StringTie\n",
    "\n",
    "1. [ ] Build up QC sets.\n",
    "\n",
    "CCBI symposium: http://talks.cam.ac.uk/talk/index/104587"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### May 14th biology talk with Fuxiang\n",
    "\n",
    "EveningComplex = ELF3 + ELF4 + LUX\n",
    "\n",
    "EC represses flowering genes, and is released on flowering\n",
    "\n",
    "ELF3([uniprot_O82804](http://www.uniprot.org/uniprot/O82804#family_and_domains)): mysterious protein with little structural information\n",
    "\n",
    "phyC --| ELF3  (mutant?)\n",
    "\n",
    "Temporal dynamics of phyC over the circadian period. (Look at RNASeq)\n",
    "\n",
    "SD allows phyC to reach a low level at night-day transition permitting a buildup of ELF3, that supposedly repress its targets during the day.\n",
    "\n",
    "preliminary observation: Genes regulated by ELF3 is upregulated at day-night transition\n",
    "\n",
    "Time points at transitions.\n",
    "\n",
    "other players: PRR37 LUX\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Cluster organisation\n",
    "\n",
    "username: feng\n",
    "hostname: sl-pw-srv01.slcu.private.cam.ac.uk\n",
    "\n",
    "Clusters:\n",
    "  1. Home cluster: \n",
    "    * HPC, normal usage. Shared bewtween Jerzy's group and Phil's group (permission groups). \n",
    "    * Storage: \n",
    "       1. Backup:      TeamPW (severl TB, to be mounted on Friday ), TeamJP.\n",
    "       1. Not backup:  synology2,  synology3 \n",
    "  1. Bigger cluster shared by SLCU: \n",
    "  1. CSCS\n",
    "\n",
    "Storage on Home\n",
    "\n",
    "CSCS Clinical School Computing Service\n",
    "\n",
    "### Data source\n",
    "\n",
    "1. RNASeq and ChIP-Seq\n",
    "  * Ming Jun\n",
    "  * Katja\n",
    "  \n",
    "### MISC\n",
    "\n",
    "Regular meeting:\n",
    "  1. Thursday preferably.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### May 15th\n",
    "\n",
    "#### RNASeq pipeline\n",
    "\n",
    "1. Data structure:\n",
    "  * RunID/SampleID/file\n",
    "  \n",
    "#### Structure Pipeline\n",
    "\n",
    "\n",
    "1. Automator script that scans through the folder, detect and merge the fastq files. (~/synology3/Software/mapRNAseq/mapRNAseq.py)\n",
    "\n",
    "1. Take the outputed raw .fastq files do the analyses and return the output to the automator script.\n",
    "\n",
    "fuxiang/script_nextgen"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Flowchart\n",
    "\n",
    "STEP 1: remove adaptors using Trimmomatic\n",
    "Package location: /home/Program_NGS_sl-pw-srv01/Trimmomatic-0.32/trimmomatic-0.32.jar\n",
    "\n",
    "Manual: [Trimmomatic](http://www.usadellab.org/cms/uploads/supplementary/Trimmomatic/TrimmomaticManual_V0.32.pdf)\n",
    "\n",
    "STEP 2: FASTQC quality control (you need to do this for the two paired-end read files separately)\n",
    "Package location: fastqc\n",
    "\n",
    "STEP 3: Map the reads to genome using HISAT\n",
    "Package location: /home/Program_NGS_sl-pw-srv01/hisat2-2.1.0/\n",
    "Location of indexed genome: /home/ref_genew/Brachypodium_Bd21_v3.1/HISAT2Index/Bdistachyon314_Bd\n",
    "\n",
    "format: indexed .fasta\n",
    "\n",
    "Use the following options: --no-mixed, --rna strandness RF, --dta\n",
    "\n",
    "also, add this \"&> align-summary.txt\" to the end of the command, to direct the HISAT output into a text file (which includes information about the mapping rate)\n",
    "\n",
    "STEP 4: Convert the SAM alignment file to BAM\n",
    "Package location: samtools\n",
    "\n",
    "STEP 5: sort the BAM file\n",
    "Package location: samtools\n",
    "\t\n",
    "STEP 6: remove duplicate reads\n",
    "Package location: /home/Program_NGS_sl-pw-srv01/picard-tools-1.103/MarkDuplicates.jar\n",
    "\n",
    "* STEP 7: index the deduplicated BAM files\n",
    "Package location: samtools\n",
    "\n",
    "* STEP 8: flagstat to get quality metrics\n",
    "Package location: samtools\n",
    "\n",
    "STEP 9: estimate genome average and normalise -- create a bedgraph file \n",
    "Package location: genomeCoverageBed\n",
    "\n",
    "STEP 10: convert bedgraph to bigwig -- to be visualised in IGV\n",
    "Package location: bedGraphToBigWig\n",
    "\n",
    "STEP 11: transcript assembly with StringTie\n",
    "Package location: /home/Program_NGS_sl-pw-srv01/stringtie-1.3.3b.Linux_x86_64/stringtie\n",
    "GTF file: /home/ref_genew/Brachypodium_Bd21_v3.1/annotation/Bdistachyon_314_v3.1.gene_exons.gtf (GFF file is in the same location too)\n",
    "\n",
    "(WWW phytozome)\n",
    "\n",
    "STEP 12: call raw counts using HTSEQ-COUNT (you'll need to use the '>' command to direct the output to a text file)\n",
    "Package location: htseq-count\n",
    "\n",
    "Final output should encompass:\n",
    "\t1. BAM file (after duplicate removal, from STEP 6)\n",
    "\t2. *Its BAI file (from STEP 7) \n",
    "\t3. HISAT align-summary.txt (STEP 3)\n",
    "\t4. StringTie output: a list of TPM values by gene (STEP 11)\n",
    "\t5. HTSEQ-COUNT output: a list of raw counts by gene (STEP 12)\n",
    "\t6. BedGraph file \n",
    "\t7. BigWig file\n",
    "\n",
    "Remainder files should be deleted to save space (especially the SAM file)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
